{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Information about this notebook\n",
    "\n",
    "This notebook is taken from the [DNA Diffusion project to load the DHSs](https://github.com/pinellolab/DNA-Diffusion/blob/main/notebooks/filter_master.ipynb \"The DNA Diffusion Project\").\n",
    "Then I have updated some of the codes:\n",
    "* I have broken down the codes into more cells such that the download of data are separate from reading information. This ensures that we do not download data from remote sites everytime we want to run the codes.\n",
    "\n",
    "* I also updated the code to download the matrix 'dat_bin_FDR01_hg38.txt.gz' using gdown. the method specified in original notebook does not work.\n",
    "\n",
    "* Next, when collecting names of cells from the matrix into a list with format celltype_encodeID I contatenated cols 'Biosample name','System' and 'DCC Library ID' as opposed to only using cols 'Biosample name' and 'DCC Library ID'. The reson being that the col 'System' contains important information like 'Cardiovascular' and includes more information related to cardiac systems."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/shakuntalabaichoo/opt/anaconda3/lib/python3.9/site-packages/pandas/core/computation/expressions.py:21: UserWarning: Pandas requires version '2.8.4' or newer of 'numexpr' (version '2.8.3' currently installed).\n",
      "  from pandas.core.computation.check import NUMEXPR_INSTALLED\n",
      "/Users/shakuntalabaichoo/opt/anaconda3/lib/python3.9/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).\n",
      "  from pandas.core import (\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import gzip\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-03-03 15:45:28--  https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz\n",
      "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.198.53\n",
      "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.198.53|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 983659424 (938M) [application/x-gzip]\n",
      "Saving to: 'hg38.fa.gz'\n",
      "\n",
      "hg38.fa.gz          100%[===================>] 938.09M  2.22MB/s    in 8m 44s  \n",
      "\n",
      "2024-03-03 15:54:14 (1.79 MB/s) - 'hg38.fa.gz' saved [983659424/983659424]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Download Reference genome\n",
    "!wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz\n",
    "!gzip -d hg38.fa.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENOME_PATH = \"hg38.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DataSource:\n",
    "    # Sourced from https://github.com/meuleman/SynthSeqs/blob/main/make_data/source.py\n",
    "\n",
    "    def __init__(self, data, filepath):\n",
    "        self.raw_data = data\n",
    "        self.filepath = filepath\n",
    "\n",
    "    @property\n",
    "    def data(self):\n",
    "        return self.raw_data\n",
    "\n",
    "\n",
    "class ReferenceGenome(DataSource):\n",
    "    \"\"\"Object for quickly loading and querying the reference genome.\"\"\"\n",
    "\n",
    "    @classmethod\n",
    "    def from_path(cls, path):\n",
    "        genome_dict = {record.id: str(record.seq).upper() for record in SeqIO.parse(path, \"fasta\")}\n",
    "        return cls(genome_dict, path)\n",
    "\n",
    "    @classmethod\n",
    "    def from_dict(cls, data_dict):\n",
    "        return cls(data_dict, filepath=None)\n",
    "\n",
    "    @property\n",
    "    def genome(self):\n",
    "        return self.data\n",
    "\n",
    "    def sequence(self, chrom, start, end):\n",
    "        chrom_sequence = self.genome[chrom]\n",
    "\n",
    "        assert end < len(chrom_sequence), (\n",
    "            f\"Sequence position bound out of range for chromosome {chrom}. \"\n",
    "            f\"{chrom} length {len(chrom_sequence)}, requested position {end}.\"\n",
    "        )\n",
    "        return chrom_sequence[start:end]\n",
    "\n",
    "\n",
    "genome = ReferenceGenome.from_path(GENOME_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download data\n",
    "!wget https://www.meuleman.org/DHS_Index_and_Vocabulary_metadata.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library order</th>\n",
       "      <th>Biosample name</th>\n",
       "      <th>Vocabulary representative</th>\n",
       "      <th>DCC Experiment ID</th>\n",
       "      <th>DCC Library ID</th>\n",
       "      <th>DCC Biosample ID</th>\n",
       "      <th>DCC File ID</th>\n",
       "      <th>Altius Aggregation ID</th>\n",
       "      <th>Altius Library ID</th>\n",
       "      <th>Altius Biosample ID</th>\n",
       "      <th>Replicate indicators</th>\n",
       "      <th>System</th>\n",
       "      <th>Subsystem</th>\n",
       "      <th>Organ</th>\n",
       "      <th>Biosample type</th>\n",
       "      <th>Biological state</th>\n",
       "      <th>Germ layer</th>\n",
       "      <th>Description</th>\n",
       "      <th>Growth stage</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Donor ID</th>\n",
       "      <th>Unique cellular condition</th>\n",
       "      <th>Used in Figure 1b</th>\n",
       "      <th>Biosample protocol</th>\n",
       "      <th>Experiment protocol</th>\n",
       "      <th>Library kit method</th>\n",
       "      <th>Library cleanup</th>\n",
       "      <th>DNaseI units/mL</th>\n",
       "      <th>Amount Nucleic Acid (ng)</th>\n",
       "      <th>Nuclei count</th>\n",
       "      <th>Protease inhibitor</th>\n",
       "      <th>Library sequencing date</th>\n",
       "      <th>Reads used</th>\n",
       "      <th>DCC SPOT score</th>\n",
       "      <th>Per-biosample peaks</th>\n",
       "      <th>DHSs in Index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GM06990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EMQ</td>\n",
       "      <td>ENCLB435ZZZ</td>\n",
       "      <td>ENCBS057ENC</td>\n",
       "      <td>ENCFF983CTQ</td>\n",
       "      <td>AG5636</td>\n",
       "      <td>LN1203</td>\n",
       "      <td>DS7748</td>\n",
       "      <td>DS7784</td>\n",
       "      <td>Hematopoietic</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>Blood</td>\n",
       "      <td>Lines</td>\n",
       "      <td>Immortalized</td>\n",
       "      <td>Mesoderm</td>\n",
       "      <td>Lymphoblastoid</td>\n",
       "      <td>Adult</td>\n",
       "      <td>41Y</td>\n",
       "      <td>F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biosample protocol</td>\n",
       "      <td>Experiment protocol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sucrose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2009-02-23</td>\n",
       "      <td>142681590.0</td>\n",
       "      <td>0.6790</td>\n",
       "      <td>83639.0</td>\n",
       "      <td>82918.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>HepG2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000ENP</td>\n",
       "      <td>ENCLB480ZZZ</td>\n",
       "      <td>ENCBS114ENC</td>\n",
       "      <td>ENCFF419JVG</td>\n",
       "      <td>AG5635</td>\n",
       "      <td>LN1207</td>\n",
       "      <td>DS7764</td>\n",
       "      <td>DS7768</td>\n",
       "      <td>Hepatic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Liver</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>Endoderm</td>\n",
       "      <td>hepatocellular carcinoma</td>\n",
       "      <td>Child</td>\n",
       "      <td>15Y</td>\n",
       "      <td>M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biosample protocol</td>\n",
       "      <td>Experiment protocol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sucrose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2009-02-23</td>\n",
       "      <td>138826342.0</td>\n",
       "      <td>0.5858</td>\n",
       "      <td>89748.0</td>\n",
       "      <td>89235.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731</th>\n",
       "      <td>732.0</td>\n",
       "      <td>fPlacenta</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR552RKI</td>\n",
       "      <td>ENCLB423VBC</td>\n",
       "      <td>ENCBS565KNL</td>\n",
       "      <td>ENCFF084UVH</td>\n",
       "      <td>AG8805</td>\n",
       "      <td>LN45072C</td>\n",
       "      <td>DS37386C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fetal Life Support</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Placenta</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Endoderm</td>\n",
       "      <td>placenta</td>\n",
       "      <td>Fetal</td>\n",
       "      <td>102D</td>\n",
       "      <td>U</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H26550</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biosample protocol</td>\n",
       "      <td>Experiment protocol</td>\n",
       "      <td>Thruplex DNA-Seq Q</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.325</td>\n",
       "      <td>1050000.0</td>\n",
       "      <td>A+Sucrose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>203699532.0</td>\n",
       "      <td>0.3869</td>\n",
       "      <td>107611.0</td>\n",
       "      <td>106022.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>732</th>\n",
       "      <td>733.0</td>\n",
       "      <td>fPlacenta</td>\n",
       "      <td>Placental / trophoblast</td>\n",
       "      <td>ENCSR552XJI</td>\n",
       "      <td>ENCLB711ZZZ</td>\n",
       "      <td>ENCBS723HLT</td>\n",
       "      <td>ENCFF593AWN</td>\n",
       "      <td>AG7450</td>\n",
       "      <td>LN45076C</td>\n",
       "      <td>DS37716C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fetal Life Support</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Placenta</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Endoderm</td>\n",
       "      <td>Placenta</td>\n",
       "      <td>Fetal</td>\n",
       "      <td>56D</td>\n",
       "      <td>U</td>\n",
       "      <td>NaN</td>\n",
       "      <td>H26598</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biosample protocol</td>\n",
       "      <td>Experiment protocol</td>\n",
       "      <td>Thruplex DNA-Seq Q</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.972</td>\n",
       "      <td>1380000.0</td>\n",
       "      <td>A+Sucrose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>206456483.0</td>\n",
       "      <td>0.4356</td>\n",
       "      <td>115898.0</td>\n",
       "      <td>114344.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>733 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     library order Biosample name Vocabulary representative DCC Experiment ID  \\\n",
       "0              1.0        GM06990                       NaN       ENCSR000EMQ   \n",
       "1              2.0          HepG2                       NaN       ENCSR000ENP   \n",
       "..             ...            ...                       ...               ...   \n",
       "731          732.0      fPlacenta                       NaN       ENCSR552RKI   \n",
       "732          733.0      fPlacenta   Placental / trophoblast       ENCSR552XJI   \n",
       "\n",
       "    DCC Library ID DCC Biosample ID  DCC File ID Altius Aggregation ID  \\\n",
       "0      ENCLB435ZZZ      ENCBS057ENC  ENCFF983CTQ                AG5636   \n",
       "1      ENCLB480ZZZ      ENCBS114ENC  ENCFF419JVG                AG5635   \n",
       "..             ...              ...          ...                   ...   \n",
       "731    ENCLB423VBC      ENCBS565KNL  ENCFF084UVH                AG8805   \n",
       "732    ENCLB711ZZZ      ENCBS723HLT  ENCFF593AWN                AG7450   \n",
       "\n",
       "    Altius Library ID Altius Biosample ID Replicate indicators  \\\n",
       "0              LN1203              DS7748               DS7784   \n",
       "1              LN1207              DS7764               DS7768   \n",
       "..                ...                 ...                  ...   \n",
       "731          LN45072C            DS37386C                  NaN   \n",
       "732          LN45076C            DS37716C                  NaN   \n",
       "\n",
       "                 System Subsystem     Organ Biosample type Biological state  \\\n",
       "0         Hematopoietic  Lymphoid     Blood          Lines     Immortalized   \n",
       "1               Hepatic       NaN     Liver         Cancer           Cancer   \n",
       "..                  ...       ...       ...            ...              ...   \n",
       "731  Fetal Life Support       NaN  Placenta        Primary          Primary   \n",
       "732  Fetal Life Support       NaN  Placenta        Primary          Primary   \n",
       "\n",
       "    Germ layer              Description  Growth stage   Age Sex Ethnicity  \\\n",
       "0     Mesoderm            Lymphoblastoid        Adult   41Y   F       NaN   \n",
       "1     Endoderm  hepatocellular carcinoma        Child   15Y   M       NaN   \n",
       "..         ...                       ...          ...   ...  ..       ...   \n",
       "731   Endoderm                  placenta        Fetal  102D   U       NaN   \n",
       "732   Endoderm                  Placenta        Fetal   56D   U       NaN   \n",
       "\n",
       "    Donor ID  Unique cellular condition  Used in Figure 1b  \\\n",
       "0        NaN                          0                NaN   \n",
       "1        NaN                          1                NaN   \n",
       "..       ...                        ...                ...   \n",
       "731  H26550                           1                NaN   \n",
       "732   H26598                          1                NaN   \n",
       "\n",
       "     Biosample protocol  Experiment protocol  Library kit method  \\\n",
       "0    Biosample protocol  Experiment protocol                 NaN   \n",
       "1    Biosample protocol  Experiment protocol                 NaN   \n",
       "..                  ...                  ...                 ...   \n",
       "731  Biosample protocol  Experiment protocol  Thruplex DNA-Seq Q   \n",
       "732  Biosample protocol  Experiment protocol  Thruplex DNA-Seq Q   \n",
       "\n",
       "    Library cleanup  DNaseI units/mL Amount Nucleic Acid (ng)  Nuclei count  \\\n",
       "0           Sucrose              NaN                       50           NaN   \n",
       "1           Sucrose              NaN                       50           NaN   \n",
       "..              ...              ...                      ...           ...   \n",
       "731             NaN              NaN                    1.325     1050000.0   \n",
       "732             NaN              NaN                    0.972     1380000.0   \n",
       "\n",
       "    Protease inhibitor Library sequencing date   Reads used  DCC SPOT score  \\\n",
       "0                  NaN              2009-02-23  142681590.0          0.6790   \n",
       "1                  NaN              2009-02-23  138826342.0          0.5858   \n",
       "..                 ...                     ...          ...             ...   \n",
       "731          A+Sucrose                     NaN  203699532.0          0.3869   \n",
       "732          A+Sucrose                     NaN  206456483.0          0.4356   \n",
       "\n",
       "     Per-biosample peaks  DHSs in Index  \n",
       "0                83639.0        82918.0  \n",
       "1                89748.0        89235.0  \n",
       "..                   ...            ...  \n",
       "731             107611.0       106022.0  \n",
       "732             115898.0       114344.0  \n",
       "\n",
       "[733 rows x 38 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Last row is empty\n",
    "DHS_Index_and_Vocabulary_metadata = pd.read_table('./DHS_Index_and_Vocabulary_metadata.tsv').iloc[:-1]\n",
    "with pd.option_context('display.max_rows', 5, 'display.max_columns', None):\n",
    "    display(DHS_Index_and_Vocabulary_metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Contains a 733 row (biosample) x 16 (component) peak presence/abscence matrix (not a binary matrix)\n",
    "# Used later to map component number within metadata dataframe and find proportion for given component\n",
    "# Downloading basis\n",
    "basis_array = requests.get(\"https://zenodo.org/record/3838751/files/2018-06-08NC16_NNDSVD_Basis.npy.gz?download=1\")\n",
    "\n",
    "with open('2018-06-08NC16_NNDSVD_Basis.npy.gz', 'wb') as f:\n",
    "    f.write(basis_array.content)\n",
    "\n",
    "!gzip -d 2018-06-08NC16_NNDSVD_Basis.npy.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Converting npy file to csv\n",
    "basis_array = np.load('2018-06-08NC16_NNDSVD_Basis.npy')\n",
    "np.savetxt(\"2018-06-08NC16_NNDSVD_Basis.csv\", basis_array, delimiter=\",\")\n",
    "\n",
    "# Creating nmf_loadings matrix from csv\n",
    "nmf_loadings = pd.read_csv('2018-06-08NC16_NNDSVD_Basis.csv', header=None)\n",
    "nmf_loadings.columns = ['C' + str(i) for i in range(1, 17)]\n",
    "\n",
    "\n",
    "# Joining metadata with component presence matrix\n",
    "DHS_Index_and_Vocabulary_metadata = pd.concat([DHS_Index_and_Vocabulary_metadata, nmf_loadings], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library order</th>\n",
       "      <th>Biosample name</th>\n",
       "      <th>Vocabulary representative</th>\n",
       "      <th>DCC Experiment ID</th>\n",
       "      <th>DCC Library ID</th>\n",
       "      <th>DCC Biosample ID</th>\n",
       "      <th>DCC File ID</th>\n",
       "      <th>Altius Aggregation ID</th>\n",
       "      <th>Altius Library ID</th>\n",
       "      <th>Altius Biosample ID</th>\n",
       "      <th>...</th>\n",
       "      <th>C7</th>\n",
       "      <th>C8</th>\n",
       "      <th>C9</th>\n",
       "      <th>C10</th>\n",
       "      <th>C11</th>\n",
       "      <th>C12</th>\n",
       "      <th>C13</th>\n",
       "      <th>C14</th>\n",
       "      <th>C15</th>\n",
       "      <th>C16</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GM06990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EMQ</td>\n",
       "      <td>ENCLB435ZZZ</td>\n",
       "      <td>ENCBS057ENC</td>\n",
       "      <td>ENCFF983CTQ</td>\n",
       "      <td>AG5636</td>\n",
       "      <td>LN1203</td>\n",
       "      <td>DS7748</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.102685</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.026774</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>HepG2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000ENP</td>\n",
       "      <td>ENCLB480ZZZ</td>\n",
       "      <td>ENCBS114ENC</td>\n",
       "      <td>ENCFF419JVG</td>\n",
       "      <td>AG5635</td>\n",
       "      <td>LN1207</td>\n",
       "      <td>DS7764</td>\n",
       "      <td>...</td>\n",
       "      <td>0.193557</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.074557</td>\n",
       "      <td>0.095928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.190564</td>\n",
       "      <td>0.416094</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>hTH1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EQC</td>\n",
       "      <td>ENCLB591ZZZ</td>\n",
       "      <td>ENCBS345AAA</td>\n",
       "      <td>ENCFF575KOF</td>\n",
       "      <td>AG5634</td>\n",
       "      <td>LN1222</td>\n",
       "      <td>DS7840</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.0</td>\n",
       "      <td>Hela</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000ENO</td>\n",
       "      <td>ENCLB479ZZZ</td>\n",
       "      <td>ENCBS890POO</td>\n",
       "      <td>ENCFF503PAE</td>\n",
       "      <td>AG4219</td>\n",
       "      <td>LN1264</td>\n",
       "      <td>DS8200</td>\n",
       "      <td>...</td>\n",
       "      <td>0.155545</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.407768</td>\n",
       "      <td>0.113676</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.420549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>CACO2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EMI</td>\n",
       "      <td>ENCLB422ZZZ</td>\n",
       "      <td>ENCBS391ENC</td>\n",
       "      <td>ENCFF977BRD</td>\n",
       "      <td>AG4218</td>\n",
       "      <td>LN1269</td>\n",
       "      <td>DS8235</td>\n",
       "      <td>...</td>\n",
       "      <td>0.131876</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.936955</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   library order Biosample name Vocabulary representative DCC Experiment ID  \\\n",
       "0            1.0        GM06990                       NaN       ENCSR000EMQ   \n",
       "1            2.0          HepG2                       NaN       ENCSR000ENP   \n",
       "2            3.0           hTH1                       NaN       ENCSR000EQC   \n",
       "3            4.0           Hela                       NaN       ENCSR000ENO   \n",
       "4            5.0          CACO2                       NaN       ENCSR000EMI   \n",
       "\n",
       "  DCC Library ID DCC Biosample ID  DCC File ID Altius Aggregation ID  \\\n",
       "0    ENCLB435ZZZ      ENCBS057ENC  ENCFF983CTQ                AG5636   \n",
       "1    ENCLB480ZZZ      ENCBS114ENC  ENCFF419JVG                AG5635   \n",
       "2    ENCLB591ZZZ      ENCBS345AAA  ENCFF575KOF                AG5634   \n",
       "3    ENCLB479ZZZ      ENCBS890POO  ENCFF503PAE                AG4219   \n",
       "4    ENCLB422ZZZ      ENCBS391ENC  ENCFF977BRD                AG4218   \n",
       "\n",
       "  Altius Library ID Altius Biosample ID  ...        C7   C8        C9  \\\n",
       "0            LN1203              DS7748  ...  0.000000  0.0  0.000000   \n",
       "1            LN1207              DS7764  ...  0.193557  0.0  0.074557   \n",
       "2            LN1222              DS7840  ...  0.000000  0.0  0.000000   \n",
       "3            LN1264              DS8200  ...  0.155545  0.0  0.000000   \n",
       "4            LN1269              DS8235  ...  0.131876  0.0  0.000000   \n",
       "\n",
       "        C10  C11  C12       C13       C14  C15       C16  \n",
       "0  0.102685  0.0  0.0  0.026774  0.000000  0.0  0.000000  \n",
       "1  0.095928  0.0  0.0  3.190564  0.416094  0.0  0.000000  \n",
       "2  0.000000  0.0  0.0  0.000000  0.000000  0.0  0.000000  \n",
       "3  0.000000  0.0  0.0  0.407768  0.113676  0.0  2.420549  \n",
       "4  0.000000  0.0  0.0  0.936955  0.000000  0.0  0.000000  \n",
       "\n",
       "[5 rows x 54 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DHS_Index_and_Vocabulary_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "COMPONENT_COLUMNS = [\n",
    "    'C1',\n",
    "    'C2',\n",
    "    'C3',\n",
    "    'C4',\n",
    "    'C5',\n",
    "    'C6',\n",
    "    'C7',\n",
    "    'C8',\n",
    "    'C9',\n",
    "    'C10',\n",
    "    'C11',\n",
    "    'C12',\n",
    "    'C13',\n",
    "    'C14',\n",
    "    'C15',\n",
    "    'C16',\n",
    "]\n",
    "\n",
    "DHS_Index_and_Vocabulary_metadata['component'] = (\n",
    "    DHS_Index_and_Vocabulary_metadata[COMPONENT_COLUMNS].idxmax(axis=1).apply(lambda x: int(x[1:]))\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating sequence metadata dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CAUTION NEXT CELL CAN TAKE UPWARDS OF 10 MINUTES TO RUN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# File loaded from drive available from below link\n",
    "mixture_array = requests.get(\"https://zenodo.org/record/3838751/files/2018-06-08NC16_NNDSVD_Mixture.npy.gz?download=1\")\n",
    "\n",
    "# Downloading mixture array that contains 3.5M x 16 matrix of peak presence/absence decomposed into 16 components\n",
    "with open('2018-06-08NC16_NNDSVD_Mixture.npy.gz', 'wb') as f:\n",
    "    f.write(mixture_array.content)\n",
    "!gzip -d 2018-06-08NC16_NNDSVD_Mixture.npy.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Turning npy file into csv\n",
    "mixture_array = np.load('2018-06-08NC16_NNDSVD_Mixture.npy').T\n",
    "np.savetxt(\"2018-06-08NC16_NNDSVD_Mixture.csv\", mixture_array, delimiter=\",\")\n",
    "\n",
    "# Creating nmf_loadings matrix from csv and renaming columns\n",
    "nmf_loadings = pd.read_csv('2018-06-08NC16_NNDSVD_Mixture.csv', header=None, names=COMPONENT_COLUMNS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download data\n",
    "!wget https://www.meuleman.org/DHS_Index_and_Vocabulary_hg38_WM20190703.txt.gz\n",
    "!gunzip -d DHS_Index_and_Vocabulary_hg38_WM20190703.txt.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/1w/414v9fyj3jg9fp1hfb7m06jh0000gn/T/ipykernel_8721/625281858.py:5: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  sequence_metadata = pd.read_table('./DHS_Index_and_Vocabulary_hg38_WM20190703.txt', sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "# Loading in DHS_Index_and_Vocabulary_metadata that contains the following information:\n",
    "# seqname, start, end, identifier, mean_signal, numsaples, summit, core_start, core_end, component\n",
    "\n",
    "# Loading sequence metadata\n",
    "sequence_metadata = pd.read_table('./DHS_Index_and_Vocabulary_hg38_WM20190703.txt', sep='\\t')\n",
    "\n",
    "# Dropping component column that contains associated tissue rather than component number (We will use the component number from DHS_Index_and_Vocabulary_metadata)\n",
    "sequence_metadata = sequence_metadata.drop(columns=['component'], axis=1)\n",
    "\n",
    "# Join metadata with component presence matrix\n",
    "df = pd.concat([sequence_metadata, nmf_loadings], axis=1, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dhs_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>DHS_width</th>\n",
       "      <th>summit</th>\n",
       "      <th>numsamples</th>\n",
       "      <th>total_signal</th>\n",
       "      <th>component</th>\n",
       "      <th>proportion</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_16140_16200_16170</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16140</td>\n",
       "      <td>16200</td>\n",
       "      <td>60</td>\n",
       "      <td>16170</td>\n",
       "      <td>1</td>\n",
       "      <td>0.129388</td>\n",
       "      <td>1</td>\n",
       "      <td>0.855153</td>\n",
       "      <td>CGGGCATCCTGTGTGCAGATACTCCCTGCTTCCTCTCTAGCCCCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_51868_52040_51970</td>\n",
       "      <td>chr1</td>\n",
       "      <td>51868</td>\n",
       "      <td>52040</td>\n",
       "      <td>172</td>\n",
       "      <td>51970</td>\n",
       "      <td>1</td>\n",
       "      <td>0.080034</td>\n",
       "      <td>7</td>\n",
       "      <td>0.973545</td>\n",
       "      <td>GGCGACCCAGCGAGACTCCGCCTCAAAAAAAAAAAAAGAAGATTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_57280_57354_57350</td>\n",
       "      <td>chr1</td>\n",
       "      <td>57280</td>\n",
       "      <td>57354</td>\n",
       "      <td>74</td>\n",
       "      <td>57350</td>\n",
       "      <td>4</td>\n",
       "      <td>1.093002</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>CTCAGTCATTCCGAACAATTCACACACTAAGATTACCCATGCTAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_66370_66482_66430</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66370</td>\n",
       "      <td>66482</td>\n",
       "      <td>112</td>\n",
       "      <td>66430</td>\n",
       "      <td>8</td>\n",
       "      <td>1.469725</td>\n",
       "      <td>3</td>\n",
       "      <td>0.332213</td>\n",
       "      <td>ATATATAAATTATATAATATAATATATATTATATAATATAATATAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_79100_79231_79150</td>\n",
       "      <td>chr1</td>\n",
       "      <td>79100</td>\n",
       "      <td>79231</td>\n",
       "      <td>131</td>\n",
       "      <td>79150</td>\n",
       "      <td>2</td>\n",
       "      <td>0.226098</td>\n",
       "      <td>7</td>\n",
       "      <td>0.501840</td>\n",
       "      <td>CATTTCTCCAAGGAGGAAATACCAGAGTCAATTCACAACCACTGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591893</th>\n",
       "      <td>chrY_56882540_56882719_56882610</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56882540</td>\n",
       "      <td>56882719</td>\n",
       "      <td>179</td>\n",
       "      <td>56882610</td>\n",
       "      <td>1</td>\n",
       "      <td>0.038079</td>\n",
       "      <td>5</td>\n",
       "      <td>0.803229</td>\n",
       "      <td>TTCTTTTTTAACACCACCACAACACTCAAGGAAATGAAAGTAGGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591894</th>\n",
       "      <td>chrY_56882864_56882980_56882930</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56882864</td>\n",
       "      <td>56882980</td>\n",
       "      <td>116</td>\n",
       "      <td>56882930</td>\n",
       "      <td>1</td>\n",
       "      <td>0.115489</td>\n",
       "      <td>5</td>\n",
       "      <td>0.742349</td>\n",
       "      <td>TCCTAATGCTATCCCTTCCCCCTCCCCCTACCCCACATTGCTCTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591895</th>\n",
       "      <td>chrY_56883733_56883960_56883830</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56883733</td>\n",
       "      <td>56883960</td>\n",
       "      <td>227</td>\n",
       "      <td>56883830</td>\n",
       "      <td>5</td>\n",
       "      <td>2.456885</td>\n",
       "      <td>7</td>\n",
       "      <td>0.559734</td>\n",
       "      <td>GGGTTGGTAATGAGGAGCCAAGGATGACTCATTTTCAGGTTGGAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591896</th>\n",
       "      <td>chrY_56884440_56884580_56884510</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56884440</td>\n",
       "      <td>56884580</td>\n",
       "      <td>140</td>\n",
       "      <td>56884510</td>\n",
       "      <td>1</td>\n",
       "      <td>0.053759</td>\n",
       "      <td>5</td>\n",
       "      <td>0.803229</td>\n",
       "      <td>AGCAAGAGCTTACAGAACAGGCAGCACCAAGTGGGAGCTGGGGGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591897</th>\n",
       "      <td>chrY_56885400_56885520_56885430</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56885400</td>\n",
       "      <td>56885520</td>\n",
       "      <td>120</td>\n",
       "      <td>56885430</td>\n",
       "      <td>1</td>\n",
       "      <td>0.062719</td>\n",
       "      <td>5</td>\n",
       "      <td>0.803229</td>\n",
       "      <td>AAAGACCATTTAATCTTGTCCTTTTAAATATTTAGTAACTCATTTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3591898 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  dhs_id   chr     start       end  DHS_width  \\\n",
       "0                 chr1_16140_16200_16170  chr1     16140     16200         60   \n",
       "1                 chr1_51868_52040_51970  chr1     51868     52040        172   \n",
       "2                 chr1_57280_57354_57350  chr1     57280     57354         74   \n",
       "3                 chr1_66370_66482_66430  chr1     66370     66482        112   \n",
       "4                 chr1_79100_79231_79150  chr1     79100     79231        131   \n",
       "...                                  ...   ...       ...       ...        ...   \n",
       "3591893  chrY_56882540_56882719_56882610  chrY  56882540  56882719        179   \n",
       "3591894  chrY_56882864_56882980_56882930  chrY  56882864  56882980        116   \n",
       "3591895  chrY_56883733_56883960_56883830  chrY  56883733  56883960        227   \n",
       "3591896  chrY_56884440_56884580_56884510  chrY  56884440  56884580        140   \n",
       "3591897  chrY_56885400_56885520_56885430  chrY  56885400  56885520        120   \n",
       "\n",
       "           summit  numsamples  total_signal  component  proportion  \\\n",
       "0           16170           1      0.129388          1    0.855153   \n",
       "1           51970           1      0.080034          7    0.973545   \n",
       "2           57350           4      1.093002          8    1.000000   \n",
       "3           66430           8      1.469725          3    0.332213   \n",
       "4           79150           2      0.226098          7    0.501840   \n",
       "...           ...         ...           ...        ...         ...   \n",
       "3591893  56882610           1      0.038079          5    0.803229   \n",
       "3591894  56882930           1      0.115489          5    0.742349   \n",
       "3591895  56883830           5      2.456885          7    0.559734   \n",
       "3591896  56884510           1      0.053759          5    0.803229   \n",
       "3591897  56885430           1      0.062719          5    0.803229   \n",
       "\n",
       "                                                  sequence  \n",
       "0        CGGGCATCCTGTGTGCAGATACTCCCTGCTTCCTCTCTAGCCCCCA...  \n",
       "1        GGCGACCCAGCGAGACTCCGCCTCAAAAAAAAAAAAAGAAGATTGA...  \n",
       "2        CTCAGTCATTCCGAACAATTCACACACTAAGATTACCCATGCTAAA...  \n",
       "3        ATATATAAATTATATAATATAATATATATTATATAATATAATATAT...  \n",
       "4        CATTTCTCCAAGGAGGAAATACCAGAGTCAATTCACAACCACTGCA...  \n",
       "...                                                    ...  \n",
       "3591893  TTCTTTTTTAACACCACCACAACACTCAAGGAAATGAAAGTAGGTA...  \n",
       "3591894  TCCTAATGCTATCCCTTCCCCCTCCCCCTACCCCACATTGCTCTTC...  \n",
       "3591895  GGGTTGGTAATGAGGAGCCAAGGATGACTCATTTTCAGGTTGGAGT...  \n",
       "3591896  AGCAAGAGCTTACAGAACAGGCAGCACCAAGTGGGAGCTGGGGGGC...  \n",
       "3591897  AAAGACCATTTAATCTTGTCCTTTTAAATATTTAGTAACTCATTTG...  \n",
       "\n",
       "[3591898 rows x 11 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Functions used to create sequence column\n",
    "def sequence_bounds(summit: int, start: int, end: int, length: int):\n",
    "    \"\"\"Calculate the sequence coordinates (bounds) for a given DHS.\n",
    "    https://github.com/meuleman/SynthSeqs/blob/main/make_data/process.py\n",
    "    \"\"\"\n",
    "    half = length // 2\n",
    "\n",
    "    if (summit - start) < half:\n",
    "        return start, start + length\n",
    "    elif (end - summit) < half:\n",
    "        return end - length, end\n",
    "\n",
    "    return summit - half, summit + half\n",
    "\n",
    "\n",
    "def add_sequence_column(df: pd.DataFrame, genome, length: int):\n",
    "    \"\"\"\n",
    "    Query the reference genome for each DHS and add the raw sequences\n",
    "    to the dataframe.\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        The dataframe of DHS annotations and NMF loadings.\n",
    "    genome : ReferenceGenome(DataSource)\n",
    "        A reference genome object to query for sequences.\n",
    "    length : int\n",
    "        Length of a DHS.\n",
    "\n",
    "    https://github.com/meuleman/SynthSeqs/blob/main/make_data/process.py\n",
    "    \"\"\"\n",
    "    seqs = []\n",
    "    for rowi, row in df.iterrows():\n",
    "        l, r = sequence_bounds(row['summit'], row['start'], row['end'], length)\n",
    "        seq = genome.sequence(row['seqname'], l, r)\n",
    "\n",
    "        seqs.append(seq)\n",
    "\n",
    "    df['sequence'] = seqs\n",
    "    return df\n",
    "\n",
    "\n",
    "# Recreating some of the columns from our original dataset\n",
    "df['component'] = df[COMPONENT_COLUMNS].idxmax(axis=1).apply(lambda x: int(x[1:]))\n",
    "df['proportion'] = df[COMPONENT_COLUMNS].max(axis=1) / df[COMPONENT_COLUMNS].sum(axis=1)\n",
    "df['total_signal'] = df['mean_signal'] * df['numsamples']\n",
    "df['proportion'] = df[COMPONENT_COLUMNS].max(axis=1) / df[COMPONENT_COLUMNS].sum(axis=1)\n",
    "df['dhs_id'] = df[['seqname', 'start', 'end', 'summit']].apply(lambda x: '_'.join(map(str, x)), axis=1)\n",
    "df['DHS_width'] = df['end'] - df['start']\n",
    "\n",
    "# Creating sequence column\n",
    "df = add_sequence_column(df, genome, 200)\n",
    "\n",
    "# Changing seqname column to chr\n",
    "df = df.rename(columns={'seqname': 'chr'})\n",
    "\n",
    "# Reordering and unselecting columns\n",
    "df = df[\n",
    "    [\n",
    "        'dhs_id',\n",
    "        'chr',\n",
    "        'start',\n",
    "        'end',\n",
    "        'DHS_width',\n",
    "        'summit',\n",
    "        'numsamples',\n",
    "        'total_signal',\n",
    "        'component',\n",
    "        'proportion',\n",
    "        'sequence',\n",
    "    ]\n",
    "]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save df as feather file so that we do not need run this everytime we are testing our code\n",
    "# It uses pyarrow; if it is not installed, we must install it, using the following\n",
    "#!pip install pyarrow\n",
    "df.to_feather('dhs_index_vocab_seq_dataset.ftr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading...\n",
      "From (original): https://drive.google.com/uc?export=download&id=16wbuNmHnwsek3USWM04nR535vPavNZka\n",
      "From (redirected): https://drive.google.com/uc?export=download&id=16wbuNmHnwsek3USWM04nR535vPavNZka&confirm=t&uuid=155233db-c762-4510-96fb-7d596fffbc37\n",
      "To: /Users/shakuntalabaichoo/Projects/ShakunJobApps/UHN/Computational Biologist Peter Munk Lab/DHS_Analysis/dat_bin_FDR01_hg38.txt.gz\n",
      "100%|████████████████████████████████████████████████████████████| 415M/415M [03:14<00:00, 2.14MB/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'dat_bin_FDR01_hg38.txt.gz'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Downloading binary peak matrix file from https://www.meuleman.org/research/dhsindex/\n",
    "\n",
    "# !wget https://drive.google.com/uc?export=download&id=1Nel7wWOWhWn40Yv7eaQFwvpMcQHBNtJ2 # does not work\n",
    "# Instead use gdown\n",
    "# Install gdown if it does not exist, using the next line\n",
    "# !pip install gdown\n",
    "\n",
    "\n",
    "import gdown\n",
    "\n",
    "url = 'https://drive.google.com/uc?export=download&id=16wbuNmHnwsek3USWM04nR535vPavNZka'\n",
    "output = 'dat_bin_FDR01_hg38.txt.gz'\n",
    "gdown.download(url, output, quiet=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extracting binary peak matrix file from zip\n",
    "!gzip -d dat_bin_FDR01_hg38.txt.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Opening file\n",
    "binary_matrix = pd.read_table('./dat_bin_FDR01_hg38.txt', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collecting names of cells into a list with format celltype_encodeID\n",
    "'''\n",
    "celltype_encodeID = [\n",
    "    row['Biosample name'] + \"_\" + row['DCC Library ID'] for _, row in DHS_Index_and_Vocabulary_metadata.iterrows()\n",
    "]\n",
    "\n",
    "# Renaming columns using celltype_encodeID list\n",
    "binary_matrix.columns = celltype_encodeID\n",
    "'''\n",
    "# I wish to add the system info as well because it contains important information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collecting names of cells into a list with format celltype_encodeID\n",
    "celltype_encodeID = [\n",
    "    row['Biosample name'] + \"_\" + row['System'].replace(\" \", \"_\") + \"_\" + row['DCC Library ID'] for _, row in DHS_Index_and_Vocabulary_metadata.iterrows()\n",
    "]\n",
    "\n",
    "# Renaming columns using celltype_encodeID list\n",
    "binary_matrix.columns = celltype_encodeID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_dataset = pd.concat([df, binary_matrix], axis=1, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dhs_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>DHS_width</th>\n",
       "      <th>summit</th>\n",
       "      <th>numsamples</th>\n",
       "      <th>total_signal</th>\n",
       "      <th>component</th>\n",
       "      <th>proportion</th>\n",
       "      <th>...</th>\n",
       "      <th>fKidney_Renal_ENCLB005SRL</th>\n",
       "      <th>fKidney_Renal_ENCLB704GMQ</th>\n",
       "      <th>fKidney_Renal_ENCLB759USM</th>\n",
       "      <th>fLung_Respiratory_ENCLB594BSZ</th>\n",
       "      <th>fKidney_Renal_ENCLB049MNH</th>\n",
       "      <th>fUmbilical_cord_Fetal_Life_Support_ENCLB771UER</th>\n",
       "      <th>fBone_femur_Musculoskeletal_ENCLB236BWV</th>\n",
       "      <th>fLiver_Hepatic_ENCLB638FEH</th>\n",
       "      <th>fPlacenta_Fetal_Life_Support_ENCLB423VBC</th>\n",
       "      <th>fPlacenta_Fetal_Life_Support_ENCLB711ZZZ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_16140_16200_16170</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16140</td>\n",
       "      <td>16200</td>\n",
       "      <td>60</td>\n",
       "      <td>16170</td>\n",
       "      <td>1</td>\n",
       "      <td>0.129388</td>\n",
       "      <td>1</td>\n",
       "      <td>0.855153</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_51868_52040_51970</td>\n",
       "      <td>chr1</td>\n",
       "      <td>51868</td>\n",
       "      <td>52040</td>\n",
       "      <td>172</td>\n",
       "      <td>51970</td>\n",
       "      <td>1</td>\n",
       "      <td>0.080034</td>\n",
       "      <td>7</td>\n",
       "      <td>0.973545</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_57280_57354_57350</td>\n",
       "      <td>chr1</td>\n",
       "      <td>57280</td>\n",
       "      <td>57354</td>\n",
       "      <td>74</td>\n",
       "      <td>57350</td>\n",
       "      <td>4</td>\n",
       "      <td>1.093002</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_66370_66482_66430</td>\n",
       "      <td>chr1</td>\n",
       "      <td>66370</td>\n",
       "      <td>66482</td>\n",
       "      <td>112</td>\n",
       "      <td>66430</td>\n",
       "      <td>8</td>\n",
       "      <td>1.469725</td>\n",
       "      <td>3</td>\n",
       "      <td>0.332213</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_79100_79231_79150</td>\n",
       "      <td>chr1</td>\n",
       "      <td>79100</td>\n",
       "      <td>79231</td>\n",
       "      <td>131</td>\n",
       "      <td>79150</td>\n",
       "      <td>2</td>\n",
       "      <td>0.226098</td>\n",
       "      <td>7</td>\n",
       "      <td>0.501840</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.12227</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591893</th>\n",
       "      <td>chrY_56882540_56882719_56882610</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56882540</td>\n",
       "      <td>56882719</td>\n",
       "      <td>179</td>\n",
       "      <td>56882610</td>\n",
       "      <td>1</td>\n",
       "      <td>0.038079</td>\n",
       "      <td>5</td>\n",
       "      <td>0.803229</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591894</th>\n",
       "      <td>chrY_56882864_56882980_56882930</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56882864</td>\n",
       "      <td>56882980</td>\n",
       "      <td>116</td>\n",
       "      <td>56882930</td>\n",
       "      <td>1</td>\n",
       "      <td>0.115489</td>\n",
       "      <td>5</td>\n",
       "      <td>0.742349</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591895</th>\n",
       "      <td>chrY_56883733_56883960_56883830</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56883733</td>\n",
       "      <td>56883960</td>\n",
       "      <td>227</td>\n",
       "      <td>56883830</td>\n",
       "      <td>5</td>\n",
       "      <td>2.456885</td>\n",
       "      <td>7</td>\n",
       "      <td>0.559734</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591896</th>\n",
       "      <td>chrY_56884440_56884580_56884510</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56884440</td>\n",
       "      <td>56884580</td>\n",
       "      <td>140</td>\n",
       "      <td>56884510</td>\n",
       "      <td>1</td>\n",
       "      <td>0.053759</td>\n",
       "      <td>5</td>\n",
       "      <td>0.803229</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591897</th>\n",
       "      <td>chrY_56885400_56885520_56885430</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56885400</td>\n",
       "      <td>56885520</td>\n",
       "      <td>120</td>\n",
       "      <td>56885430</td>\n",
       "      <td>1</td>\n",
       "      <td>0.062719</td>\n",
       "      <td>5</td>\n",
       "      <td>0.803229</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3591898 rows × 744 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  dhs_id   chr     start       end  DHS_width  \\\n",
       "0                 chr1_16140_16200_16170  chr1     16140     16200         60   \n",
       "1                 chr1_51868_52040_51970  chr1     51868     52040        172   \n",
       "2                 chr1_57280_57354_57350  chr1     57280     57354         74   \n",
       "3                 chr1_66370_66482_66430  chr1     66370     66482        112   \n",
       "4                 chr1_79100_79231_79150  chr1     79100     79231        131   \n",
       "...                                  ...   ...       ...       ...        ...   \n",
       "3591893  chrY_56882540_56882719_56882610  chrY  56882540  56882719        179   \n",
       "3591894  chrY_56882864_56882980_56882930  chrY  56882864  56882980        116   \n",
       "3591895  chrY_56883733_56883960_56883830  chrY  56883733  56883960        227   \n",
       "3591896  chrY_56884440_56884580_56884510  chrY  56884440  56884580        140   \n",
       "3591897  chrY_56885400_56885520_56885430  chrY  56885400  56885520        120   \n",
       "\n",
       "           summit  numsamples  total_signal  component  proportion  ...  \\\n",
       "0           16170           1      0.129388          1    0.855153  ...   \n",
       "1           51970           1      0.080034          7    0.973545  ...   \n",
       "2           57350           4      1.093002          8    1.000000  ...   \n",
       "3           66430           8      1.469725          3    0.332213  ...   \n",
       "4           79150           2      0.226098          7    0.501840  ...   \n",
       "...           ...         ...           ...        ...         ...  ...   \n",
       "3591893  56882610           1      0.038079          5    0.803229  ...   \n",
       "3591894  56882930           1      0.115489          5    0.742349  ...   \n",
       "3591895  56883830           5      2.456885          7    0.559734  ...   \n",
       "3591896  56884510           1      0.053759          5    0.803229  ...   \n",
       "3591897  56885430           1      0.062719          5    0.803229  ...   \n",
       "\n",
       "        fKidney_Renal_ENCLB005SRL  fKidney_Renal_ENCLB704GMQ  \\\n",
       "0                             0.0                        0.0   \n",
       "1                             0.0                        0.0   \n",
       "2                             0.0                        0.0   \n",
       "3                             0.0                        0.0   \n",
       "4                             0.0                        0.0   \n",
       "...                           ...                        ...   \n",
       "3591893                       0.0                        0.0   \n",
       "3591894                       0.0                        0.0   \n",
       "3591895                       0.0                        0.0   \n",
       "3591896                       0.0                        0.0   \n",
       "3591897                       0.0                        0.0   \n",
       "\n",
       "         fKidney_Renal_ENCLB759USM  fLung_Respiratory_ENCLB594BSZ  \\\n",
       "0                              0.0                        0.00000   \n",
       "1                              0.0                        0.00000   \n",
       "2                              0.0                        0.00000   \n",
       "3                              0.0                        0.00000   \n",
       "4                              0.0                        0.12227   \n",
       "...                            ...                            ...   \n",
       "3591893                        0.0                        0.00000   \n",
       "3591894                        0.0                        0.00000   \n",
       "3591895                        0.0                        0.00000   \n",
       "3591896                        0.0                        0.00000   \n",
       "3591897                        0.0                        0.00000   \n",
       "\n",
       "         fKidney_Renal_ENCLB049MNH  \\\n",
       "0                              0.0   \n",
       "1                              0.0   \n",
       "2                              0.0   \n",
       "3                              0.0   \n",
       "4                              0.0   \n",
       "...                            ...   \n",
       "3591893                        0.0   \n",
       "3591894                        0.0   \n",
       "3591895                        0.0   \n",
       "3591896                        0.0   \n",
       "3591897                        0.0   \n",
       "\n",
       "         fUmbilical_cord_Fetal_Life_Support_ENCLB771UER  \\\n",
       "0                                                   0.0   \n",
       "1                                                   0.0   \n",
       "2                                                   0.0   \n",
       "3                                                   0.0   \n",
       "4                                                   0.0   \n",
       "...                                                 ...   \n",
       "3591893                                             0.0   \n",
       "3591894                                             0.0   \n",
       "3591895                                             0.0   \n",
       "3591896                                             0.0   \n",
       "3591897                                             0.0   \n",
       "\n",
       "         fBone_femur_Musculoskeletal_ENCLB236BWV  fLiver_Hepatic_ENCLB638FEH  \\\n",
       "0                                            0.0                         0.0   \n",
       "1                                            0.0                         0.0   \n",
       "2                                            0.0                         0.0   \n",
       "3                                            0.0                         0.0   \n",
       "4                                            0.0                         0.0   \n",
       "...                                          ...                         ...   \n",
       "3591893                                      0.0                         0.0   \n",
       "3591894                                      0.0                         0.0   \n",
       "3591895                                      0.0                         0.0   \n",
       "3591896                                      0.0                         0.0   \n",
       "3591897                                      0.0                         0.0   \n",
       "\n",
       "         fPlacenta_Fetal_Life_Support_ENCLB423VBC  \\\n",
       "0                                             0.0   \n",
       "1                                             0.0   \n",
       "2                                             0.0   \n",
       "3                                             0.0   \n",
       "4                                             0.0   \n",
       "...                                           ...   \n",
       "3591893                                       0.0   \n",
       "3591894                                       0.0   \n",
       "3591895                                       0.0   \n",
       "3591896                                       0.0   \n",
       "3591897                                       0.0   \n",
       "\n",
       "         fPlacenta_Fetal_Life_Support_ENCLB711ZZZ  \n",
       "0                                             0.0  \n",
       "1                                             0.0  \n",
       "2                                             0.0  \n",
       "3                                             0.0  \n",
       "4                                             0.0  \n",
       "...                                           ...  \n",
       "3591893                                       0.0  \n",
       "3591894                                       0.0  \n",
       "3591895                                       0.0  \n",
       "3591896                                       0.0  \n",
       "3591897                                       0.0  \n",
       "\n",
       "[3591898 rows x 744 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "master_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save as feather file - it uses pyarrow; if it does not exist, we must install it, using the following\n",
    "#!pip install pyarrow\n",
    "\n",
    "master_dataset.to_feather('master_dataset.ftr')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "vscode": {
   "interpreter": {
    "hash": "e4d3eab395c0ad02fe5331a19eec9b67cdc045e607aca948639107f1caddcb62"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
