{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ce7b2640",
   "metadata": {},
   "source": [
    "# DHS Data preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0db7a80",
   "metadata": {},
   "source": [
    "This notebook is used to prepare the DHS data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2e4a7374",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/elwazi_mru/anaconda3/lib/python3.11/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).\n",
      "  from pandas.core import (\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "from io import BytesIO\n",
    "from gzip import GzipFile"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ad2b00c",
   "metadata": {},
   "source": [
    "## 1.0 Download the DHS Vocabulary metadata and read into a dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c7294346",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define URLs and paths for the metadadata\n",
    "dhs_index_vocab_metadata_url = \"http://www.meuleman.org/DHS_Index_and_Vocabulary_metadata.tsv\"\n",
    "dhs_index_vocab_metadata_path = \"./Problem2_DHS_data/DHS_Index_and_Vocabulary_metadata.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "db12ca02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download the tsv file\n",
    "response = requests.get(dhs_index_vocab_metadata_url)\n",
    "with open(dhs_index_vocab_metadata_path, 'wb') as file:\n",
    "    file.write(response.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "afcb4de7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library order</th>\n",
       "      <th>Biosample name</th>\n",
       "      <th>Vocabulary representative</th>\n",
       "      <th>DCC Experiment ID</th>\n",
       "      <th>DCC Library ID</th>\n",
       "      <th>DCC Biosample ID</th>\n",
       "      <th>DCC File ID</th>\n",
       "      <th>Altius Aggregation ID</th>\n",
       "      <th>Altius Library ID</th>\n",
       "      <th>Altius Biosample ID</th>\n",
       "      <th>...</th>\n",
       "      <th>Library cleanup</th>\n",
       "      <th>DNaseI units/mL</th>\n",
       "      <th>Amount Nucleic Acid (ng)</th>\n",
       "      <th>Nuclei count</th>\n",
       "      <th>Protease inhibitor</th>\n",
       "      <th>Library sequencing date</th>\n",
       "      <th>Reads used</th>\n",
       "      <th>DCC SPOT score</th>\n",
       "      <th>Per-biosample peaks</th>\n",
       "      <th>DHSs in Index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>GM06990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EMQ</td>\n",
       "      <td>ENCLB435ZZZ</td>\n",
       "      <td>ENCBS057ENC</td>\n",
       "      <td>ENCFF983CTQ</td>\n",
       "      <td>AG5636</td>\n",
       "      <td>LN1203</td>\n",
       "      <td>DS7748</td>\n",
       "      <td>...</td>\n",
       "      <td>Sucrose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2009-02-23</td>\n",
       "      <td>142681590.0</td>\n",
       "      <td>0.6790</td>\n",
       "      <td>83639.0</td>\n",
       "      <td>82918.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>HepG2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000ENP</td>\n",
       "      <td>ENCLB480ZZZ</td>\n",
       "      <td>ENCBS114ENC</td>\n",
       "      <td>ENCFF419JVG</td>\n",
       "      <td>AG5635</td>\n",
       "      <td>LN1207</td>\n",
       "      <td>DS7764</td>\n",
       "      <td>...</td>\n",
       "      <td>Sucrose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2009-02-23</td>\n",
       "      <td>138826342.0</td>\n",
       "      <td>0.5858</td>\n",
       "      <td>89748.0</td>\n",
       "      <td>89235.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>hTH1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EQC</td>\n",
       "      <td>ENCLB591ZZZ</td>\n",
       "      <td>ENCBS345AAA</td>\n",
       "      <td>ENCFF575KOF</td>\n",
       "      <td>AG5634</td>\n",
       "      <td>LN1222</td>\n",
       "      <td>DS7840</td>\n",
       "      <td>...</td>\n",
       "      <td>Sucrose</td>\n",
       "      <td>6.0</td>\n",
       "      <td>534.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2007-06-06</td>\n",
       "      <td>149158633.0</td>\n",
       "      <td>0.6470</td>\n",
       "      <td>94360.0</td>\n",
       "      <td>93665.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.0</td>\n",
       "      <td>Hela</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000ENO</td>\n",
       "      <td>ENCLB479ZZZ</td>\n",
       "      <td>ENCBS890POO</td>\n",
       "      <td>ENCFF503PAE</td>\n",
       "      <td>AG4219</td>\n",
       "      <td>LN1264</td>\n",
       "      <td>DS8200</td>\n",
       "      <td>...</td>\n",
       "      <td>new Sucrose</td>\n",
       "      <td>4.0</td>\n",
       "      <td>50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2007-08-24</td>\n",
       "      <td>23372724.0</td>\n",
       "      <td>0.6444</td>\n",
       "      <td>59098.0</td>\n",
       "      <td>59024.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>CACO2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EMI</td>\n",
       "      <td>ENCLB422ZZZ</td>\n",
       "      <td>ENCBS391ENC</td>\n",
       "      <td>ENCFF977BRD</td>\n",
       "      <td>AG4218</td>\n",
       "      <td>LN1269</td>\n",
       "      <td>DS8235</td>\n",
       "      <td>...</td>\n",
       "      <td>Sucrose</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2007-09-05</td>\n",
       "      <td>22760059.0</td>\n",
       "      <td>0.7190</td>\n",
       "      <td>29894.0</td>\n",
       "      <td>29724.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>729</th>\n",
       "      <td>730.0</td>\n",
       "      <td>fBone_femur</td>\n",
       "      <td>Musculoskeletal</td>\n",
       "      <td>ENCSR805XIF</td>\n",
       "      <td>ENCLB236BWV</td>\n",
       "      <td>ENCBS337FPV</td>\n",
       "      <td>ENCFF604WIO</td>\n",
       "      <td>AG7442</td>\n",
       "      <td>LN45038B</td>\n",
       "      <td>DS36206B</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.8</td>\n",
       "      <td>1050000.0</td>\n",
       "      <td>A+Sucrose</td>\n",
       "      <td>2017-02-17</td>\n",
       "      <td>252066174.0</td>\n",
       "      <td>0.5823</td>\n",
       "      <td>146918.0</td>\n",
       "      <td>145356.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>730</th>\n",
       "      <td>731.0</td>\n",
       "      <td>fLiver</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR562FNN</td>\n",
       "      <td>ENCLB638FEH</td>\n",
       "      <td>ENCBS275VNY</td>\n",
       "      <td>ENCFF795ZXN</td>\n",
       "      <td>AG7443</td>\n",
       "      <td>LN45070C</td>\n",
       "      <td>DS37372C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.48</td>\n",
       "      <td>2140000.0</td>\n",
       "      <td>A+Sucrose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>190541422.0</td>\n",
       "      <td>0.3703</td>\n",
       "      <td>76639.0</td>\n",
       "      <td>75369.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731</th>\n",
       "      <td>732.0</td>\n",
       "      <td>fPlacenta</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR552RKI</td>\n",
       "      <td>ENCLB423VBC</td>\n",
       "      <td>ENCBS565KNL</td>\n",
       "      <td>ENCFF084UVH</td>\n",
       "      <td>AG8805</td>\n",
       "      <td>LN45072C</td>\n",
       "      <td>DS37386C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.325</td>\n",
       "      <td>1050000.0</td>\n",
       "      <td>A+Sucrose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>203699532.0</td>\n",
       "      <td>0.3869</td>\n",
       "      <td>107611.0</td>\n",
       "      <td>106022.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>732</th>\n",
       "      <td>733.0</td>\n",
       "      <td>fPlacenta</td>\n",
       "      <td>Placental / trophoblast</td>\n",
       "      <td>ENCSR552XJI</td>\n",
       "      <td>ENCLB711ZZZ</td>\n",
       "      <td>ENCBS723HLT</td>\n",
       "      <td>ENCFF593AWN</td>\n",
       "      <td>AG7450</td>\n",
       "      <td>LN45076C</td>\n",
       "      <td>DS37716C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.972</td>\n",
       "      <td>1380000.0</td>\n",
       "      <td>A+Sucrose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>206456483.0</td>\n",
       "      <td>0.4356</td>\n",
       "      <td>115898.0</td>\n",
       "      <td>114344.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>733</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>734 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     library order Biosample name Vocabulary representative DCC Experiment ID  \\\n",
       "0              1.0        GM06990                       NaN       ENCSR000EMQ   \n",
       "1              2.0          HepG2                       NaN       ENCSR000ENP   \n",
       "2              3.0           hTH1                       NaN       ENCSR000EQC   \n",
       "3              4.0           Hela                       NaN       ENCSR000ENO   \n",
       "4              5.0          CACO2                       NaN       ENCSR000EMI   \n",
       "..             ...            ...                       ...               ...   \n",
       "729          730.0    fBone_femur           Musculoskeletal       ENCSR805XIF   \n",
       "730          731.0         fLiver                       NaN       ENCSR562FNN   \n",
       "731          732.0      fPlacenta                       NaN       ENCSR552RKI   \n",
       "732          733.0      fPlacenta   Placental / trophoblast       ENCSR552XJI   \n",
       "733            NaN            NaN                       NaN               NaN   \n",
       "\n",
       "    DCC Library ID DCC Biosample ID  DCC File ID Altius Aggregation ID  \\\n",
       "0      ENCLB435ZZZ      ENCBS057ENC  ENCFF983CTQ                AG5636   \n",
       "1      ENCLB480ZZZ      ENCBS114ENC  ENCFF419JVG                AG5635   \n",
       "2      ENCLB591ZZZ      ENCBS345AAA  ENCFF575KOF                AG5634   \n",
       "3      ENCLB479ZZZ      ENCBS890POO  ENCFF503PAE                AG4219   \n",
       "4      ENCLB422ZZZ      ENCBS391ENC  ENCFF977BRD                AG4218   \n",
       "..             ...              ...          ...                   ...   \n",
       "729    ENCLB236BWV      ENCBS337FPV  ENCFF604WIO                AG7442   \n",
       "730    ENCLB638FEH      ENCBS275VNY  ENCFF795ZXN                AG7443   \n",
       "731    ENCLB423VBC      ENCBS565KNL  ENCFF084UVH                AG8805   \n",
       "732    ENCLB711ZZZ      ENCBS723HLT  ENCFF593AWN                AG7450   \n",
       "733            NaN              NaN          NaN                   NaN   \n",
       "\n",
       "    Altius Library ID Altius Biosample ID  ... Library cleanup  \\\n",
       "0              LN1203              DS7748  ...         Sucrose   \n",
       "1              LN1207              DS7764  ...         Sucrose   \n",
       "2              LN1222              DS7840  ...         Sucrose   \n",
       "3              LN1264              DS8200  ...     new Sucrose   \n",
       "4              LN1269              DS8235  ...         Sucrose   \n",
       "..                ...                 ...  ...             ...   \n",
       "729          LN45038B            DS36206B  ...             NaN   \n",
       "730          LN45070C            DS37372C  ...             NaN   \n",
       "731          LN45072C            DS37386C  ...             NaN   \n",
       "732          LN45076C            DS37716C  ...             NaN   \n",
       "733               NaN                 NaN  ...             NaN   \n",
       "\n",
       "    DNaseI units/mL Amount Nucleic Acid (ng) Nuclei count Protease inhibitor  \\\n",
       "0               NaN                       50          NaN                NaN   \n",
       "1               NaN                       50          NaN                NaN   \n",
       "2               6.0                    534.9          NaN                NaN   \n",
       "3               4.0                       50          NaN                NaN   \n",
       "4               8.0                        1          NaN                NaN   \n",
       "..              ...                      ...          ...                ...   \n",
       "729             NaN                      8.8    1050000.0          A+Sucrose   \n",
       "730             NaN                     4.48    2140000.0          A+Sucrose   \n",
       "731             NaN                    1.325    1050000.0          A+Sucrose   \n",
       "732             NaN                    0.972    1380000.0          A+Sucrose   \n",
       "733             NaN                      NaN          NaN                NaN   \n",
       "\n",
       "    Library sequencing date   Reads used DCC SPOT score Per-biosample peaks  \\\n",
       "0                2009-02-23  142681590.0         0.6790             83639.0   \n",
       "1                2009-02-23  138826342.0         0.5858             89748.0   \n",
       "2                2007-06-06  149158633.0         0.6470             94360.0   \n",
       "3                2007-08-24   23372724.0         0.6444             59098.0   \n",
       "4                2007-09-05   22760059.0         0.7190             29894.0   \n",
       "..                      ...          ...            ...                 ...   \n",
       "729              2017-02-17  252066174.0         0.5823            146918.0   \n",
       "730                     NaN  190541422.0         0.3703             76639.0   \n",
       "731                     NaN  203699532.0         0.3869            107611.0   \n",
       "732                     NaN  206456483.0         0.4356            115898.0   \n",
       "733                     NaN          NaN            NaN                 NaN   \n",
       "\n",
       "    DHSs in Index  \n",
       "0         82918.0  \n",
       "1         89235.0  \n",
       "2         93665.0  \n",
       "3         59024.0  \n",
       "4         29724.0  \n",
       "..            ...  \n",
       "729      145356.0  \n",
       "730       75369.0  \n",
       "731      106022.0  \n",
       "732      114344.0  \n",
       "733           NaN  \n",
       "\n",
       "[734 rows x 38 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the DHS vocabulary metadata into a dataframe\n",
    "data_df_vocab = pd.read_csv(dhs_index_vocab_metadata_path, sep=\"\\t\")\n",
    "data_df_vocab"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a9b7993",
   "metadata": {},
   "source": [
    "## 2.0 Download the DHS Index and Vocabulary data and read into a dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "42689290",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define URLs and paths for the dadata\n",
    "dhs_index_vocab_url = \"http://www.meuleman.org/DHS_Index_and_Vocabulary_hg38_WM20190703.txt.gz\"\n",
    "dhs_index_vocab_path = \"./Problem2_DHS_data/DHS_Index_and_Vocabulary_hg38_WM20190703.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "763d3f6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download the gz file\n",
    "response = requests.get(dhs_index_vocab_url)\n",
    "with open(dhs_index_vocab_path, 'wb') as file:\n",
    "    file.write(response.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b5119034",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_82210/3105284989.py:3: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data_df_index_Vocab = pd.read_csv(gzfile, sep=\"\\t\")\n"
     ]
    }
   ],
   "source": [
    "# Read the compressed tab-limited txt file directly into a dataframe\n",
    "with GzipFile(dhs_index_vocab_path) as gzfile:\n",
    "    data_df_index_Vocab = pd.read_csv(gzfile, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6066371c",
   "metadata": {},
   "source": [
    "# 3.0 Filter out rows to retain most significant DHSs.\n",
    "## Using Z-Scores\n",
    "We will use a statistical approach, Z-scores, which measure the number of standard deviations an element is from the mean. This method is particularly useful if we are interested in finding DHSs that are significantly higher (or lower) in numsamples and mean_signal relative to the overall distribution:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ba6ea450",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate Z-scores for 'numsamples' and 'mean_signal'\n",
    "data_df_index_Vocab['numsamples_z'] = (data_df_index_Vocab['numsamples'] - data_df_index_Vocab['numsamples'].mean()) / data_df_index_Vocab['numsamples'].std()\n",
    "data_df_index_Vocab['mean_signal_z'] = (data_df_index_Vocab['mean_signal'] - data_df_index_Vocab['mean_signal'].mean()) / data_df_index_Vocab['mean_signal'].std()\n",
    "\n",
    "# Filter based on Z-scores (e.g., keep entries with Z-scores above 1)\n",
    "data_df_index_Vocab_short = data_df_index_Vocab[\n",
    "    (data_df_index_Vocab['numsamples_z'] > 1) &\n",
    "    (data_df_index_Vocab['mean_signal_z'] > 1)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "49cb5252",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>identifier</th>\n",
       "      <th>mean_signal</th>\n",
       "      <th>numsamples</th>\n",
       "      <th>summit</th>\n",
       "      <th>core_start</th>\n",
       "      <th>core_end</th>\n",
       "      <th>component</th>\n",
       "      <th>numsamples_z</th>\n",
       "      <th>mean_signal_z</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr1</td>\n",
       "      <td>181400</td>\n",
       "      <td>181564</td>\n",
       "      <td>1.100655</td>\n",
       "      <td>1.712263</td>\n",
       "      <td>540</td>\n",
       "      <td>181490</td>\n",
       "      <td>181450.0</td>\n",
       "      <td>181490.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>7.762808</td>\n",
       "      <td>1.141472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629160</td>\n",
       "      <td>629310</td>\n",
       "      <td>1.102262</td>\n",
       "      <td>40.454979</td>\n",
       "      <td>281</td>\n",
       "      <td>629230</td>\n",
       "      <td>629210.0</td>\n",
       "      <td>629270.0</td>\n",
       "      <td>Stromal A</td>\n",
       "      <td>3.887120</td>\n",
       "      <td>42.390107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629520</td>\n",
       "      <td>629596</td>\n",
       "      <td>1.102264</td>\n",
       "      <td>5.152820</td>\n",
       "      <td>187</td>\n",
       "      <td>629590</td>\n",
       "      <td>629530.0</td>\n",
       "      <td>629590.0</td>\n",
       "      <td>Stromal A</td>\n",
       "      <td>2.480500</td>\n",
       "      <td>4.804567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629870</td>\n",
       "      <td>630020</td>\n",
       "      <td>1.102264</td>\n",
       "      <td>146.867423</td>\n",
       "      <td>730</td>\n",
       "      <td>629930</td>\n",
       "      <td>629930.0</td>\n",
       "      <td>629970.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>10.605976</td>\n",
       "      <td>155.685418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630181</td>\n",
       "      <td>630319</td>\n",
       "      <td>1.102266</td>\n",
       "      <td>41.684094</td>\n",
       "      <td>143</td>\n",
       "      <td>630270</td>\n",
       "      <td>630210.0</td>\n",
       "      <td>630310.0</td>\n",
       "      <td>Stromal A</td>\n",
       "      <td>1.822082</td>\n",
       "      <td>43.698722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590512</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19354160</td>\n",
       "      <td>19354375</td>\n",
       "      <td>Y.40428</td>\n",
       "      <td>1.702447</td>\n",
       "      <td>101</td>\n",
       "      <td>19354270</td>\n",
       "      <td>19354230.0</td>\n",
       "      <td>19354300.0</td>\n",
       "      <td>Stromal B</td>\n",
       "      <td>1.193592</td>\n",
       "      <td>1.131021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590609</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19567050</td>\n",
       "      <td>19567360</td>\n",
       "      <td>Y.407629</td>\n",
       "      <td>4.718289</td>\n",
       "      <td>395</td>\n",
       "      <td>19567210</td>\n",
       "      <td>19567170.0</td>\n",
       "      <td>19567293.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>5.593021</td>\n",
       "      <td>4.341931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590719</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19744660</td>\n",
       "      <td>19745060</td>\n",
       "      <td>Y.41039</td>\n",
       "      <td>2.006189</td>\n",
       "      <td>379</td>\n",
       "      <td>19744810</td>\n",
       "      <td>19744730.0</td>\n",
       "      <td>19744990.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>5.353596</td>\n",
       "      <td>1.454409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590982</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20575532</td>\n",
       "      <td>20575800</td>\n",
       "      <td>Y.423413</td>\n",
       "      <td>5.157244</td>\n",
       "      <td>396</td>\n",
       "      <td>20575670</td>\n",
       "      <td>20575630.0</td>\n",
       "      <td>20575710.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>5.607985</td>\n",
       "      <td>4.809277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591718</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56734848</td>\n",
       "      <td>56734900</td>\n",
       "      <td>Y.99198</td>\n",
       "      <td>4.689927</td>\n",
       "      <td>384</td>\n",
       "      <td>56734850</td>\n",
       "      <td>56734850.0</td>\n",
       "      <td>56734850.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>5.428417</td>\n",
       "      <td>4.311734</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>93906 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname     start       end identifier  mean_signal  numsamples  \\\n",
       "20         chr1    181400    181564   1.100655     1.712263         540   \n",
       "45         chr1    629160    629310   1.102262    40.454979         281   \n",
       "47         chr1    629520    629596   1.102264     5.152820         187   \n",
       "48         chr1    629870    630020   1.102264   146.867423         730   \n",
       "50         chr1    630181    630319   1.102266    41.684094         143   \n",
       "...         ...       ...       ...        ...          ...         ...   \n",
       "3590512    chrY  19354160  19354375    Y.40428     1.702447         101   \n",
       "3590609    chrY  19567050  19567360   Y.407629     4.718289         395   \n",
       "3590719    chrY  19744660  19745060    Y.41039     2.006189         379   \n",
       "3590982    chrY  20575532  20575800   Y.423413     5.157244         396   \n",
       "3591718    chrY  56734848  56734900    Y.99198     4.689927         384   \n",
       "\n",
       "           summit  core_start    core_end         component  numsamples_z  \\\n",
       "20         181490    181450.0    181490.0  Tissue invariant      7.762808   \n",
       "45         629230    629210.0    629270.0         Stromal A      3.887120   \n",
       "47         629590    629530.0    629590.0         Stromal A      2.480500   \n",
       "48         629930    629930.0    629970.0  Tissue invariant     10.605976   \n",
       "50         630270    630210.0    630310.0         Stromal A      1.822082   \n",
       "...           ...         ...         ...               ...           ...   \n",
       "3590512  19354270  19354230.0  19354300.0         Stromal B      1.193592   \n",
       "3590609  19567210  19567170.0  19567293.0  Tissue invariant      5.593021   \n",
       "3590719  19744810  19744730.0  19744990.0  Tissue invariant      5.353596   \n",
       "3590982  20575670  20575630.0  20575710.0  Tissue invariant      5.607985   \n",
       "3591718  56734850  56734850.0  56734850.0  Tissue invariant      5.428417   \n",
       "\n",
       "         mean_signal_z  \n",
       "20            1.141472  \n",
       "45           42.390107  \n",
       "47            4.804567  \n",
       "48          155.685418  \n",
       "50           43.698722  \n",
       "...                ...  \n",
       "3590512       1.131021  \n",
       "3590609       4.341931  \n",
       "3590719       1.454409  \n",
       "3590982       4.809277  \n",
       "3591718       4.311734  \n",
       "\n",
       "[93906 rows x 12 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_df_index_Vocab_short"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fafc102",
   "metadata": {},
   "source": [
    "# 4.0 Define Cardiac Tissue-Specific DHSs\n",
    "First, define \"heart-related\" versus \"not-heart-related\" \n",
    "Heart-Related: Sequences derived from tissues such as cardiomyocytes, cardiac fibroblasts, endothelial cells from heart tissues, etc.\n",
    "Not-Heart-Related: Sequences derived from all other non-cardiac tissues."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "278713a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library order</th>\n",
       "      <th>Biosample name</th>\n",
       "      <th>Vocabulary representative</th>\n",
       "      <th>DCC Experiment ID</th>\n",
       "      <th>DCC Library ID</th>\n",
       "      <th>DCC Biosample ID</th>\n",
       "      <th>DCC File ID</th>\n",
       "      <th>Altius Aggregation ID</th>\n",
       "      <th>Altius Library ID</th>\n",
       "      <th>Altius Biosample ID</th>\n",
       "      <th>...</th>\n",
       "      <th>Library cleanup</th>\n",
       "      <th>DNaseI units/mL</th>\n",
       "      <th>Amount Nucleic Acid (ng)</th>\n",
       "      <th>Nuclei count</th>\n",
       "      <th>Protease inhibitor</th>\n",
       "      <th>Library sequencing date</th>\n",
       "      <th>Reads used</th>\n",
       "      <th>DCC SPOT score</th>\n",
       "      <th>Per-biosample peaks</th>\n",
       "      <th>DHSs in Index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18.0</td>\n",
       "      <td>HUVEC</td>\n",
       "      <td>Vascular / endothelial</td>\n",
       "      <td>ENCSR000EOQ</td>\n",
       "      <td>ENCLB533ZZZ</td>\n",
       "      <td>ENCBS112ENC</td>\n",
       "      <td>ENCFF011FHJ</td>\n",
       "      <td>AG4209</td>\n",
       "      <td>LN1812</td>\n",
       "      <td>DS10060</td>\n",
       "      <td>...</td>\n",
       "      <td>Sucrose</td>\n",
       "      <td>60.0</td>\n",
       "      <td>24.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2008-12-10</td>\n",
       "      <td>383483914.0</td>\n",
       "      <td>0.4182</td>\n",
       "      <td>175498.0</td>\n",
       "      <td>174822.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>42.0</td>\n",
       "      <td>H7_hESC_T14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EMY</td>\n",
       "      <td>ENCLB447ZZZ</td>\n",
       "      <td>ENCBS532SEX</td>\n",
       "      <td>ENCFF575JJO</td>\n",
       "      <td>AG4189</td>\n",
       "      <td>LN2329</td>\n",
       "      <td>DS11814</td>\n",
       "      <td>...</td>\n",
       "      <td>minElute</td>\n",
       "      <td>120.0</td>\n",
       "      <td>2.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2009-08-07</td>\n",
       "      <td>28331633.0</td>\n",
       "      <td>0.3604</td>\n",
       "      <td>53849.0</td>\n",
       "      <td>53784.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>49.0</td>\n",
       "      <td>H7_hESC_T5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EMW</td>\n",
       "      <td>ENCLB444ZZZ</td>\n",
       "      <td>ENCBS430FQD</td>\n",
       "      <td>ENCFF358BDI</td>\n",
       "      <td>AG4183</td>\n",
       "      <td>LN2375</td>\n",
       "      <td>DS11953</td>\n",
       "      <td>...</td>\n",
       "      <td>minElute</td>\n",
       "      <td>60.0</td>\n",
       "      <td>8.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2009-08-24</td>\n",
       "      <td>27961281.0</td>\n",
       "      <td>0.3867</td>\n",
       "      <td>61948.0</td>\n",
       "      <td>61860.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>54.0</td>\n",
       "      <td>H7_hESC_T14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000EMY</td>\n",
       "      <td>ENCLB448ZZZ</td>\n",
       "      <td>ENCBS294AAA</td>\n",
       "      <td>ENCFF936NEG</td>\n",
       "      <td>AG4178</td>\n",
       "      <td>LN2425</td>\n",
       "      <td>DS12147</td>\n",
       "      <td>...</td>\n",
       "      <td>minElute</td>\n",
       "      <td>120.0</td>\n",
       "      <td>10.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2009-09-02</td>\n",
       "      <td>26153486.0</td>\n",
       "      <td>0.3755</td>\n",
       "      <td>52350.0</td>\n",
       "      <td>52293.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>76.0</td>\n",
       "      <td>HCF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR000ENH</td>\n",
       "      <td>ENCLB464ZZZ</td>\n",
       "      <td>ENCBS756HNZ</td>\n",
       "      <td>ENCFF996CVV</td>\n",
       "      <td>AG4159</td>\n",
       "      <td>LN2540</td>\n",
       "      <td>DS12491</td>\n",
       "      <td>...</td>\n",
       "      <td>minElute</td>\n",
       "      <td>120.0</td>\n",
       "      <td>14.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2009-09-30</td>\n",
       "      <td>27918106.0</td>\n",
       "      <td>0.6044</td>\n",
       "      <td>80246.0</td>\n",
       "      <td>80178.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>688.0</td>\n",
       "      <td>fHeart</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>ENCSR749BWV</td>\n",
       "      <td>ENCLB491BID</td>\n",
       "      <td>ENCBS320XNN</td>\n",
       "      <td>ENCFF981VXY</td>\n",
       "      <td>AG4019</td>\n",
       "      <td>LN31529A</td>\n",
       "      <td>DS37473A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>2400000.0</td>\n",
       "      <td>A+Sucrose</td>\n",
       "      <td>2015-06-01</td>\n",
       "      <td>218768197.0</td>\n",
       "      <td>0.4300</td>\n",
       "      <td>141622.0</td>\n",
       "      <td>141121.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688</th>\n",
       "      <td>689.0</td>\n",
       "      <td>fLeftAtrium</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>ENCSR412NMI</td>\n",
       "      <td>ENCLB226FNM</td>\n",
       "      <td>ENCBS982OCU</td>\n",
       "      <td>ENCFF741AUL</td>\n",
       "      <td>AG6213</td>\n",
       "      <td>LN31530A</td>\n",
       "      <td>DS37460A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>529000.0</td>\n",
       "      <td>A+Sucrose</td>\n",
       "      <td>2015-06-01</td>\n",
       "      <td>212960129.0</td>\n",
       "      <td>0.3496</td>\n",
       "      <td>108967.0</td>\n",
       "      <td>108488.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691</th>\n",
       "      <td>692.0</td>\n",
       "      <td>fLeftVentricle</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>ENCSR501FWC</td>\n",
       "      <td>ENCLB699DLF</td>\n",
       "      <td>ENCBS936JZF</td>\n",
       "      <td>ENCFF957QNS</td>\n",
       "      <td>AG3860</td>\n",
       "      <td>LN31860S</td>\n",
       "      <td>DS37292S</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.9</td>\n",
       "      <td>2320000.0</td>\n",
       "      <td>A+Sucrose</td>\n",
       "      <td>2015-07-01</td>\n",
       "      <td>203158085.0</td>\n",
       "      <td>0.3587</td>\n",
       "      <td>134029.0</td>\n",
       "      <td>133445.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>702</th>\n",
       "      <td>703.0</td>\n",
       "      <td>fHeartFibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR856NXV</td>\n",
       "      <td>ENCLB584PEG</td>\n",
       "      <td>ENCBS385GBH</td>\n",
       "      <td>ENCFF409ZKB</td>\n",
       "      <td>AG6217</td>\n",
       "      <td>LN34648A</td>\n",
       "      <td>DS39481A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.336</td>\n",
       "      <td>900000.0</td>\n",
       "      <td>PBS+A</td>\n",
       "      <td>2016-01-06</td>\n",
       "      <td>199930072.0</td>\n",
       "      <td>0.5550</td>\n",
       "      <td>200548.0</td>\n",
       "      <td>200221.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>705.0</td>\n",
       "      <td>fLeftVentricle</td>\n",
       "      <td>Cardiac</td>\n",
       "      <td>ENCSR946DXB</td>\n",
       "      <td>ENCLB231DPY</td>\n",
       "      <td>ENCBS743EKM</td>\n",
       "      <td>ENCFF743DVR</td>\n",
       "      <td>AG6219</td>\n",
       "      <td>LN35747A</td>\n",
       "      <td>DS44708A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.9515</td>\n",
       "      <td>2760000.0</td>\n",
       "      <td>A+Sucrose</td>\n",
       "      <td>2016-04-01</td>\n",
       "      <td>191669306.0</td>\n",
       "      <td>0.3734</td>\n",
       "      <td>130087.0</td>\n",
       "      <td>129365.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     library order     Biosample name Vocabulary representative  \\\n",
       "17            18.0              HUVEC    Vascular / endothelial   \n",
       "41            42.0        H7_hESC_T14                       NaN   \n",
       "48            49.0         H7_hESC_T5                       NaN   \n",
       "53            54.0        H7_hESC_T14                       NaN   \n",
       "75            76.0                HCF                       NaN   \n",
       "..             ...                ...                       ...   \n",
       "687          688.0             fHeart                   Cardiac   \n",
       "688          689.0        fLeftAtrium                   Cardiac   \n",
       "691          692.0     fLeftVentricle                   Cardiac   \n",
       "702          703.0  fHeartFibroblasts                       NaN   \n",
       "704          705.0     fLeftVentricle                   Cardiac   \n",
       "\n",
       "    DCC Experiment ID DCC Library ID DCC Biosample ID  DCC File ID  \\\n",
       "17        ENCSR000EOQ    ENCLB533ZZZ      ENCBS112ENC  ENCFF011FHJ   \n",
       "41        ENCSR000EMY    ENCLB447ZZZ      ENCBS532SEX  ENCFF575JJO   \n",
       "48        ENCSR000EMW    ENCLB444ZZZ      ENCBS430FQD  ENCFF358BDI   \n",
       "53        ENCSR000EMY    ENCLB448ZZZ      ENCBS294AAA  ENCFF936NEG   \n",
       "75        ENCSR000ENH    ENCLB464ZZZ      ENCBS756HNZ  ENCFF996CVV   \n",
       "..                ...            ...              ...          ...   \n",
       "687       ENCSR749BWV    ENCLB491BID      ENCBS320XNN  ENCFF981VXY   \n",
       "688       ENCSR412NMI    ENCLB226FNM      ENCBS982OCU  ENCFF741AUL   \n",
       "691       ENCSR501FWC    ENCLB699DLF      ENCBS936JZF  ENCFF957QNS   \n",
       "702       ENCSR856NXV    ENCLB584PEG      ENCBS385GBH  ENCFF409ZKB   \n",
       "704       ENCSR946DXB    ENCLB231DPY      ENCBS743EKM  ENCFF743DVR   \n",
       "\n",
       "    Altius Aggregation ID Altius Library ID Altius Biosample ID  ...  \\\n",
       "17                 AG4209            LN1812             DS10060  ...   \n",
       "41                 AG4189            LN2329             DS11814  ...   \n",
       "48                 AG4183            LN2375             DS11953  ...   \n",
       "53                 AG4178            LN2425             DS12147  ...   \n",
       "75                 AG4159            LN2540             DS12491  ...   \n",
       "..                    ...               ...                 ...  ...   \n",
       "687                AG4019          LN31529A            DS37473A  ...   \n",
       "688                AG6213          LN31530A            DS37460A  ...   \n",
       "691                AG3860          LN31860S            DS37292S  ...   \n",
       "702                AG6217          LN34648A            DS39481A  ...   \n",
       "704                AG6219          LN35747A            DS44708A  ...   \n",
       "\n",
       "    Library cleanup DNaseI units/mL Amount Nucleic Acid (ng) Nuclei count  \\\n",
       "17          Sucrose            60.0                     24.6          NaN   \n",
       "41         minElute           120.0                      2.2          NaN   \n",
       "48         minElute            60.0                      8.7          NaN   \n",
       "53         minElute           120.0                     10.3          NaN   \n",
       "75         minElute           120.0                     14.3          NaN   \n",
       "..              ...             ...                      ...          ...   \n",
       "687             NaN             NaN                        3    2400000.0   \n",
       "688             NaN             NaN                        3     529000.0   \n",
       "691             NaN             NaN                      1.9    2320000.0   \n",
       "702             NaN             NaN                    3.336     900000.0   \n",
       "704             NaN             NaN                   1.9515    2760000.0   \n",
       "\n",
       "    Protease inhibitor Library sequencing date   Reads used DCC SPOT score  \\\n",
       "17                 NaN              2008-12-10  383483914.0         0.4182   \n",
       "41                 NaN              2009-08-07   28331633.0         0.3604   \n",
       "48                 NaN              2009-08-24   27961281.0         0.3867   \n",
       "53                 NaN              2009-09-02   26153486.0         0.3755   \n",
       "75                 NaN              2009-09-30   27918106.0         0.6044   \n",
       "..                 ...                     ...          ...            ...   \n",
       "687          A+Sucrose              2015-06-01  218768197.0         0.4300   \n",
       "688          A+Sucrose              2015-06-01  212960129.0         0.3496   \n",
       "691          A+Sucrose              2015-07-01  203158085.0         0.3587   \n",
       "702              PBS+A              2016-01-06  199930072.0         0.5550   \n",
       "704          A+Sucrose              2016-04-01  191669306.0         0.3734   \n",
       "\n",
       "    Per-biosample peaks DHSs in Index  \n",
       "17             175498.0      174822.0  \n",
       "41              53849.0       53784.0  \n",
       "48              61948.0       61860.0  \n",
       "53              52350.0       52293.0  \n",
       "75              80246.0       80178.0  \n",
       "..                  ...           ...  \n",
       "687            141622.0      141121.0  \n",
       "688            108967.0      108488.0  \n",
       "691            134029.0      133445.0  \n",
       "702            200548.0      200221.0  \n",
       "704            130087.0      129365.0  \n",
       "\n",
       "[61 rows x 38 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter heart-related vocab\n",
    "heart_related_vocab = data_df_vocab[\n",
    "    (data_df_vocab['System'] == \"Cardiovascular\") |\n",
    "    (data_df_vocab['Subsystem'].isin([\"Cardiac Muscle\", \"Cardiac\"])) |\n",
    "    (data_df_vocab['Organ'] == \"Heart\")\n",
    "]\n",
    "heart_related_vocab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "35697cf1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_82210/161916716.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  data_df_index_Vocab_short['heart_related'] = data_df_index_Vocab_short['component'].isin(heart_related_vocab['Vocabulary representative']).astype(int)\n"
     ]
    }
   ],
   "source": [
    "# Create a binary label for heart_related\n",
    "data_df_index_Vocab_short['heart_related'] = data_df_index_Vocab_short['component'].isin(heart_related_vocab['Vocabulary representative']).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e6b751a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>identifier</th>\n",
       "      <th>mean_signal</th>\n",
       "      <th>numsamples</th>\n",
       "      <th>summit</th>\n",
       "      <th>core_start</th>\n",
       "      <th>core_end</th>\n",
       "      <th>component</th>\n",
       "      <th>numsamples_z</th>\n",
       "      <th>mean_signal_z</th>\n",
       "      <th>heart_related</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr1</td>\n",
       "      <td>181400</td>\n",
       "      <td>181564</td>\n",
       "      <td>1.100655</td>\n",
       "      <td>1.712263</td>\n",
       "      <td>540</td>\n",
       "      <td>181490</td>\n",
       "      <td>181450.0</td>\n",
       "      <td>181490.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>7.762808</td>\n",
       "      <td>1.141472</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629160</td>\n",
       "      <td>629310</td>\n",
       "      <td>1.102262</td>\n",
       "      <td>40.454979</td>\n",
       "      <td>281</td>\n",
       "      <td>629230</td>\n",
       "      <td>629210.0</td>\n",
       "      <td>629270.0</td>\n",
       "      <td>Stromal A</td>\n",
       "      <td>3.887120</td>\n",
       "      <td>42.390107</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629520</td>\n",
       "      <td>629596</td>\n",
       "      <td>1.102264</td>\n",
       "      <td>5.152820</td>\n",
       "      <td>187</td>\n",
       "      <td>629590</td>\n",
       "      <td>629530.0</td>\n",
       "      <td>629590.0</td>\n",
       "      <td>Stromal A</td>\n",
       "      <td>2.480500</td>\n",
       "      <td>4.804567</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629870</td>\n",
       "      <td>630020</td>\n",
       "      <td>1.102264</td>\n",
       "      <td>146.867423</td>\n",
       "      <td>730</td>\n",
       "      <td>629930</td>\n",
       "      <td>629930.0</td>\n",
       "      <td>629970.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>10.605976</td>\n",
       "      <td>155.685418</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630181</td>\n",
       "      <td>630319</td>\n",
       "      <td>1.102266</td>\n",
       "      <td>41.684094</td>\n",
       "      <td>143</td>\n",
       "      <td>630270</td>\n",
       "      <td>630210.0</td>\n",
       "      <td>630310.0</td>\n",
       "      <td>Stromal A</td>\n",
       "      <td>1.822082</td>\n",
       "      <td>43.698722</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590512</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19354160</td>\n",
       "      <td>19354375</td>\n",
       "      <td>Y.40428</td>\n",
       "      <td>1.702447</td>\n",
       "      <td>101</td>\n",
       "      <td>19354270</td>\n",
       "      <td>19354230.0</td>\n",
       "      <td>19354300.0</td>\n",
       "      <td>Stromal B</td>\n",
       "      <td>1.193592</td>\n",
       "      <td>1.131021</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590609</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19567050</td>\n",
       "      <td>19567360</td>\n",
       "      <td>Y.407629</td>\n",
       "      <td>4.718289</td>\n",
       "      <td>395</td>\n",
       "      <td>19567210</td>\n",
       "      <td>19567170.0</td>\n",
       "      <td>19567293.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>5.593021</td>\n",
       "      <td>4.341931</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590719</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19744660</td>\n",
       "      <td>19745060</td>\n",
       "      <td>Y.41039</td>\n",
       "      <td>2.006189</td>\n",
       "      <td>379</td>\n",
       "      <td>19744810</td>\n",
       "      <td>19744730.0</td>\n",
       "      <td>19744990.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>5.353596</td>\n",
       "      <td>1.454409</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590982</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20575532</td>\n",
       "      <td>20575800</td>\n",
       "      <td>Y.423413</td>\n",
       "      <td>5.157244</td>\n",
       "      <td>396</td>\n",
       "      <td>20575670</td>\n",
       "      <td>20575630.0</td>\n",
       "      <td>20575710.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>5.607985</td>\n",
       "      <td>4.809277</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591718</th>\n",
       "      <td>chrY</td>\n",
       "      <td>56734848</td>\n",
       "      <td>56734900</td>\n",
       "      <td>Y.99198</td>\n",
       "      <td>4.689927</td>\n",
       "      <td>384</td>\n",
       "      <td>56734850</td>\n",
       "      <td>56734850.0</td>\n",
       "      <td>56734850.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>5.428417</td>\n",
       "      <td>4.311734</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>93906 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname     start       end identifier  mean_signal  numsamples  \\\n",
       "20         chr1    181400    181564   1.100655     1.712263         540   \n",
       "45         chr1    629160    629310   1.102262    40.454979         281   \n",
       "47         chr1    629520    629596   1.102264     5.152820         187   \n",
       "48         chr1    629870    630020   1.102264   146.867423         730   \n",
       "50         chr1    630181    630319   1.102266    41.684094         143   \n",
       "...         ...       ...       ...        ...          ...         ...   \n",
       "3590512    chrY  19354160  19354375    Y.40428     1.702447         101   \n",
       "3590609    chrY  19567050  19567360   Y.407629     4.718289         395   \n",
       "3590719    chrY  19744660  19745060    Y.41039     2.006189         379   \n",
       "3590982    chrY  20575532  20575800   Y.423413     5.157244         396   \n",
       "3591718    chrY  56734848  56734900    Y.99198     4.689927         384   \n",
       "\n",
       "           summit  core_start    core_end         component  numsamples_z  \\\n",
       "20         181490    181450.0    181490.0  Tissue invariant      7.762808   \n",
       "45         629230    629210.0    629270.0         Stromal A      3.887120   \n",
       "47         629590    629530.0    629590.0         Stromal A      2.480500   \n",
       "48         629930    629930.0    629970.0  Tissue invariant     10.605976   \n",
       "50         630270    630210.0    630310.0         Stromal A      1.822082   \n",
       "...           ...         ...         ...               ...           ...   \n",
       "3590512  19354270  19354230.0  19354300.0         Stromal B      1.193592   \n",
       "3590609  19567210  19567170.0  19567293.0  Tissue invariant      5.593021   \n",
       "3590719  19744810  19744730.0  19744990.0  Tissue invariant      5.353596   \n",
       "3590982  20575670  20575630.0  20575710.0  Tissue invariant      5.607985   \n",
       "3591718  56734850  56734850.0  56734850.0  Tissue invariant      5.428417   \n",
       "\n",
       "         mean_signal_z  heart_related  \n",
       "20            1.141472              0  \n",
       "45           42.390107              0  \n",
       "47            4.804567              0  \n",
       "48          155.685418              0  \n",
       "50           43.698722              0  \n",
       "...                ...            ...  \n",
       "3590512       1.131021              1  \n",
       "3590609       4.341931              0  \n",
       "3590719       1.454409              0  \n",
       "3590982       4.809277              0  \n",
       "3591718       4.311734              0  \n",
       "\n",
       "[93906 rows x 13 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_df_index_Vocab_short"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "291213a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_82210/577006869.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  data_df_index_Vocab_short['dhs_id'] = data_df_index_Vocab_short[['seqname', 'start', 'end', 'summit']].apply(lambda x: '_'.join(map(str, x)), axis=1)\n"
     ]
    }
   ],
   "source": [
    "# create dhs_id\n",
    "data_df_index_Vocab_short['dhs_id'] = data_df_index_Vocab_short[['seqname', 'start', 'end', 'summit']].apply(lambda x: '_'.join(map(str, x)), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5e256177",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['seqname', 'start', 'end', 'identifier', 'mean_signal', 'numsamples',\n",
       "       'summit', 'core_start', 'core_end', 'component', 'numsamples_z',\n",
       "       'mean_signal_z', 'heart_related', 'dhs_id'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_df_index_Vocab_short.columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "34a2c7ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dhs_id</th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>identifier</th>\n",
       "      <th>mean_signal</th>\n",
       "      <th>numsamples</th>\n",
       "      <th>summit</th>\n",
       "      <th>core_start</th>\n",
       "      <th>core_end</th>\n",
       "      <th>component</th>\n",
       "      <th>numsamples_z</th>\n",
       "      <th>mean_signal_z</th>\n",
       "      <th>heart_related</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr1_181400_181564_181490</td>\n",
       "      <td>chr1</td>\n",
       "      <td>181400</td>\n",
       "      <td>181564</td>\n",
       "      <td>1.100655</td>\n",
       "      <td>1.712263</td>\n",
       "      <td>540</td>\n",
       "      <td>181490</td>\n",
       "      <td>181450.0</td>\n",
       "      <td>181490.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>7.762808</td>\n",
       "      <td>1.141472</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>chr1_629160_629310_629230</td>\n",
       "      <td>chr1</td>\n",
       "      <td>629160</td>\n",
       "      <td>629310</td>\n",
       "      <td>1.102262</td>\n",
       "      <td>40.454979</td>\n",
       "      <td>281</td>\n",
       "      <td>629230</td>\n",
       "      <td>629210.0</td>\n",
       "      <td>629270.0</td>\n",
       "      <td>Stromal A</td>\n",
       "      <td>3.887120</td>\n",
       "      <td>42.390107</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>chr1_629520_629596_629590</td>\n",
       "      <td>chr1</td>\n",
       "      <td>629520</td>\n",
       "      <td>629596</td>\n",
       "      <td>1.102264</td>\n",
       "      <td>5.152820</td>\n",
       "      <td>187</td>\n",
       "      <td>629590</td>\n",
       "      <td>629530.0</td>\n",
       "      <td>629590.0</td>\n",
       "      <td>Stromal A</td>\n",
       "      <td>2.480500</td>\n",
       "      <td>4.804567</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>chr1_629870_630020_629930</td>\n",
       "      <td>chr1</td>\n",
       "      <td>629870</td>\n",
       "      <td>630020</td>\n",
       "      <td>1.102264</td>\n",
       "      <td>146.867423</td>\n",
       "      <td>730</td>\n",
       "      <td>629930</td>\n",
       "      <td>629930.0</td>\n",
       "      <td>629970.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>10.605976</td>\n",
       "      <td>155.685418</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr1_630181_630319_630270</td>\n",
       "      <td>chr1</td>\n",
       "      <td>630181</td>\n",
       "      <td>630319</td>\n",
       "      <td>1.102266</td>\n",
       "      <td>41.684094</td>\n",
       "      <td>143</td>\n",
       "      <td>630270</td>\n",
       "      <td>630210.0</td>\n",
       "      <td>630310.0</td>\n",
       "      <td>Stromal A</td>\n",
       "      <td>1.822082</td>\n",
       "      <td>43.698722</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590512</th>\n",
       "      <td>chrY_19354160_19354375_19354270</td>\n",
       "      <td>chrY</td>\n",
       "      <td>19354160</td>\n",
       "      <td>19354375</td>\n",
       "      <td>Y.40428</td>\n",
       "      <td>1.702447</td>\n",
       "      <td>101</td>\n",
       "      <td>19354270</td>\n",
       "      <td>19354230.0</td>\n",
       "      <td>19354300.0</td>\n",
       "      <td>Stromal B</td>\n",
       "      <td>1.193592</td>\n",
       "      <td>1.131021</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590609</th>\n",
       "      <td>chrY_19567050_19567360_19567210</td>\n",
       "      <td>chrY</td>\n",
       "      <td>19567050</td>\n",
       "      <td>19567360</td>\n",
       "      <td>Y.407629</td>\n",
       "      <td>4.718289</td>\n",
       "      <td>395</td>\n",
       "      <td>19567210</td>\n",
       "      <td>19567170.0</td>\n",
       "      <td>19567293.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>5.593021</td>\n",
       "      <td>4.341931</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590719</th>\n",
       "      <td>chrY_19744660_19745060_19744810</td>\n",
       "      <td>chrY</td>\n",
       "      <td>19744660</td>\n",
       "      <td>19745060</td>\n",
       "      <td>Y.41039</td>\n",
       "      <td>2.006189</td>\n",
       "      <td>379</td>\n",
       "      <td>19744810</td>\n",
       "      <td>19744730.0</td>\n",
       "      <td>19744990.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>5.353596</td>\n",
       "      <td>1.454409</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590982</th>\n",
       "      <td>chrY_20575532_20575800_20575670</td>\n",
       "      <td>chrY</td>\n",
       "      <td>20575532</td>\n",
       "      <td>20575800</td>\n",
       "      <td>Y.423413</td>\n",
       "      <td>5.157244</td>\n",
       "      <td>396</td>\n",
       "      <td>20575670</td>\n",
       "      <td>20575630.0</td>\n",
       "      <td>20575710.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>5.607985</td>\n",
       "      <td>4.809277</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591718</th>\n",
       "      <td>chrY_56734848_56734900_56734850</td>\n",
       "      <td>chrY</td>\n",
       "      <td>56734848</td>\n",
       "      <td>56734900</td>\n",
       "      <td>Y.99198</td>\n",
       "      <td>4.689927</td>\n",
       "      <td>384</td>\n",
       "      <td>56734850</td>\n",
       "      <td>56734850.0</td>\n",
       "      <td>56734850.0</td>\n",
       "      <td>Tissue invariant</td>\n",
       "      <td>5.428417</td>\n",
       "      <td>4.311734</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>93906 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  dhs_id seqname     start       end  \\\n",
       "20             chr1_181400_181564_181490    chr1    181400    181564   \n",
       "45             chr1_629160_629310_629230    chr1    629160    629310   \n",
       "47             chr1_629520_629596_629590    chr1    629520    629596   \n",
       "48             chr1_629870_630020_629930    chr1    629870    630020   \n",
       "50             chr1_630181_630319_630270    chr1    630181    630319   \n",
       "...                                  ...     ...       ...       ...   \n",
       "3590512  chrY_19354160_19354375_19354270    chrY  19354160  19354375   \n",
       "3590609  chrY_19567050_19567360_19567210    chrY  19567050  19567360   \n",
       "3590719  chrY_19744660_19745060_19744810    chrY  19744660  19745060   \n",
       "3590982  chrY_20575532_20575800_20575670    chrY  20575532  20575800   \n",
       "3591718  chrY_56734848_56734900_56734850    chrY  56734848  56734900   \n",
       "\n",
       "        identifier  mean_signal  numsamples    summit  core_start    core_end  \\\n",
       "20        1.100655     1.712263         540    181490    181450.0    181490.0   \n",
       "45        1.102262    40.454979         281    629230    629210.0    629270.0   \n",
       "47        1.102264     5.152820         187    629590    629530.0    629590.0   \n",
       "48        1.102264   146.867423         730    629930    629930.0    629970.0   \n",
       "50        1.102266    41.684094         143    630270    630210.0    630310.0   \n",
       "...            ...          ...         ...       ...         ...         ...   \n",
       "3590512    Y.40428     1.702447         101  19354270  19354230.0  19354300.0   \n",
       "3590609   Y.407629     4.718289         395  19567210  19567170.0  19567293.0   \n",
       "3590719    Y.41039     2.006189         379  19744810  19744730.0  19744990.0   \n",
       "3590982   Y.423413     5.157244         396  20575670  20575630.0  20575710.0   \n",
       "3591718    Y.99198     4.689927         384  56734850  56734850.0  56734850.0   \n",
       "\n",
       "                component  numsamples_z  mean_signal_z  heart_related  \n",
       "20       Tissue invariant      7.762808       1.141472              0  \n",
       "45              Stromal A      3.887120      42.390107              0  \n",
       "47              Stromal A      2.480500       4.804567              0  \n",
       "48       Tissue invariant     10.605976     155.685418              0  \n",
       "50              Stromal A      1.822082      43.698722              0  \n",
       "...                   ...           ...            ...            ...  \n",
       "3590512         Stromal B      1.193592       1.131021              1  \n",
       "3590609  Tissue invariant      5.593021       4.341931              0  \n",
       "3590719  Tissue invariant      5.353596       1.454409              0  \n",
       "3590982  Tissue invariant      5.607985       4.809277              0  \n",
       "3591718  Tissue invariant      5.428417       4.311734              0  \n",
       "\n",
       "[93906 rows x 14 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_df_index_Vocab_short=data_df_index_Vocab_short[['dhs_id','seqname', 'start', 'end', 'identifier', 'mean_signal', 'numsamples',\n",
    "       'summit', 'core_start', 'core_end', 'component', 'numsamples_z',\n",
    "       'mean_signal_z', 'heart_related']]\n",
    "data_df_index_Vocab_short"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "4d3b5a76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the filtered dataframe\n",
    "data_df_index_Vocab_short.to_csv(\"./Problem2_DHS_data/heart_related_dhs.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2435b80b",
   "metadata": {},
   "source": [
    "# 5.0 Extract the DHS sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "00d2b104",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       dhs_id  \\\n",
      "20  chr1_181400_181564_181490   \n",
      "45  chr1_629160_629310_629230   \n",
      "47  chr1_629520_629596_629590   \n",
      "48  chr1_629870_630020_629930   \n",
      "50  chr1_630181_630319_630270   \n",
      "\n",
      "                                             sequence  heart_related  \n",
      "20  cgcccaggggaggaggcgtggcgcaggcgcagagaggcgcgccgtg...              0  \n",
      "45  CACAAACATTATTATAATAAACACCCTCACCACTACAATCTTCCTA...              0  \n",
      "47  ccatccctgagaatccaaaattctccgtgccacctatcacacccca...              0  \n",
      "48  CAATATACTCTCCGGACAATGAACCATAACCAATACCACCAATCAA...              0  \n",
      "50  ACTCCTCAATTACCCACATAGGATGAATAACAGCAGTTCTACCGTA...              0  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_82210/177440908.py:24: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  data_df_index_Vocab_short['sequence'] = data_df_index_Vocab_short.apply(extract_sequence, axis=1, genome=genome)\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pandas as pd\n",
    "\n",
    "# Assuming data_df_index_Vocab_short is your DataFrame with 'seqname', 'start', 'end', and 'heart_related' columns\n",
    "# Load the FASTA file\n",
    "fasta_file = \"./hg38.fa\"  # Replace with your FASTA file path\n",
    "genome = SeqIO.to_dict(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "\n",
    "# Function to extract sequence\n",
    "def extract_sequence(row, genome):\n",
    "    chrom = row['seqname']\n",
    "    start = row['start']\n",
    "    end = row['end']\n",
    "    # The chromosome name in the FASTA file might start with 'chr', you might need to adjust this line\n",
    "    seq = genome.get(chrom, None)\n",
    "    if seq:\n",
    "        return str(seq.seq[start-1:end])  # -1 because genomic positions are 1-based and Python is 0-based\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "# Apply the function to each row in the DataFrame\n",
    "data_df_index_Vocab_short['sequence'] = data_df_index_Vocab_short.apply(extract_sequence, axis=1, genome=genome)\n",
    "\n",
    "# Now, create a new DataFrame with sequences and the 'heart_related' column\n",
    "sequences_df = data_df_index_Vocab_short[['dhs_id','sequence', 'heart_related']].dropna()\n",
    "\n",
    "# Check the first few entries\n",
    "print(sequences_df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "7982bb2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_df.to_csv(\"../DNABERT/examples/dhs2/heart_specific_dhs_sequences.tsv\",sep='\\t', index=False, header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4452ba8a",
   "metadata": {},
   "source": [
    "## Using DNABERT\n",
    "### 1. Finetuning the DNABERT kmer 6 model\n",
    "In case we want to finetune the pretrained DNABERT model, we must have the Train and Test datasets such that we can finetune the pre-trained DNABERT model using kmers=6 using the train data and test using the test dataset.\n",
    "\n",
    "### 2. Using the finetuned pre-trained DNABERT model\n",
    "In case we wish to use the pre-trained finetuned model using kmers=6, we must download it from the [drive specified in the DNABERT repo](https://drive.google.com/drive/folders/15wFcukTv3ecPw9_25dcOv-bZmj-8d_-6). That drive contains a folder named <b>dna_model</b> with 5 files. We must download the folder dna_model and save it in our folder ```../DNABERT/examples/dhs2/ft/6/```.\n",
    "\n",
    "If using the finetuned pretrained DNABERT model, we will use the whole dataset to make predictions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f66c1209",
   "metadata": {},
   "source": [
    "# 6.0 Prepare the data by extracting 6-mers as we will use the DNABERT model trained with k-mers=6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "365c7a49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split into 6-mers (Python doesn't have a direct equivalent to R's sapply function)\n",
    "def split_into_6mers(sequence):\n",
    "    return ' '.join([sequence[i:i+6] for i in range(len(sequence) - 5)])\n",
    "\n",
    "# Assuming you have a DataFrame 'sequences_df' with the extracted sequences\n",
    "sequences_df['sequence_6mers'] = sequences_df['sequence'].apply(split_into_6mers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "12dc661e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_df1=sequences_df[['dhs_id','sequence_6mers', 'heart_related']].rename(columns={'heart_related': 'label'})\n",
    "# Write the DataFrame to a TSV file\n",
    "sequences_df1.to_csv(\"../DNABERT/examples/dhs2/ft/6/heart_specific_dhs_6mers_with_dhs_ids.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "c9c0afe2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>cgccca gcccag cccagg ccaggg cagggg agggga gggg...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>CACAAA ACAAAC CAAACA AAACAT AACATT ACATTA CATT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>ccatcc catccc atccct tccctg ccctga cctgag ctga...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>CAATAT AATATA ATATAC TATACT ATACTC TACTCT ACTC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>ACTCCT CTCCTC TCCTCA CCTCAA CTCAAT TCAATT CAAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590512</th>\n",
       "      <td>tgtgag gtgagc tgagct gagctg agctgt gctgtt ctgt...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590609</th>\n",
       "      <td>CTAGAA TAGAAC AGAACG GAACGT AACGTT ACGTTG CGTT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590719</th>\n",
       "      <td>TAGCTG AGCTGC GCTGCT CTGCTT TGCTTA GCTTAC CTTA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3590982</th>\n",
       "      <td>GGGCCC GGCCCC GCCCCG CCCCGC CCCGCC CCGCCC CGCC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3591718</th>\n",
       "      <td>tcgatt cgattg gattgg attgga ttggaa tggaat ggaa...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>93906 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  sequence  label\n",
       "20       cgccca gcccag cccagg ccaggg cagggg agggga gggg...      0\n",
       "45       CACAAA ACAAAC CAAACA AAACAT AACATT ACATTA CATT...      0\n",
       "47       ccatcc catccc atccct tccctg ccctga cctgag ctga...      0\n",
       "48       CAATAT AATATA ATATAC TATACT ATACTC TACTCT ACTC...      0\n",
       "50       ACTCCT CTCCTC TCCTCA CCTCAA CTCAAT TCAATT CAAT...      0\n",
       "...                                                    ...    ...\n",
       "3590512  tgtgag gtgagc tgagct gagctg agctgt gctgtt ctgt...      1\n",
       "3590609  CTAGAA TAGAAC AGAACG GAACGT AACGTT ACGTTG CGTT...      0\n",
       "3590719  TAGCTG AGCTGC GCTGCT CTGCTT TGCTTA GCTTAC CTTA...      0\n",
       "3590982  GGGCCC GGCCCC GCCCCG CCCCGC CCCGCC CCGCCC CGCC...      0\n",
       "3591718  tcgatt cgattg gattgg attgga ttggaa tggaat ggaa...      0\n",
       "\n",
       "[93906 rows x 2 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select the 'sequence_6mers' and 'heart_related' columns, and rename 'heart_related' to 'label'\n",
    "sequences_df_final = sequences_df[['sequence_6mers', 'heart_related']].rename(columns={'heart_related': 'label', 'sequence_6mers':'sequence'})\n",
    "sequences_df_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e3c72d88",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the DataFrame to a TSV file\n",
    "sequences_df_final.to_csv(\"../DNABERT/examples/dhs2/ft/6/heart_specific_dhs_6mers.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9120cce4",
   "metadata": {},
   "source": [
    "### Next we have to create the train, test datasets and save."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "680751c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "# Split the data into training and test sets\n",
    "train_df, test_df = train_test_split(sequences_df_final, test_size=0.2, random_state=42)\n",
    "\n",
    "# Save the training and testing sets to .tsv files\n",
    "train_df.to_csv('../DNABERT/examples/dhs2/ft/6/heart_specific_dhs_6mers_train.tsv', sep='\\t', index=False, header=False)\n",
    "test_df.to_csv('../DNABERT/examples/dhs2/ft/6/heart_specific_dhs_6mers_test.tsv', sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0c19b14",
   "metadata": {},
   "source": [
    "## Important Note\n",
    "When running prediction, the data that has to be used for prediction must be named <b>dev.tsv</b> in folder:\n",
    "../DNABERT/examples/dhs2/ft/6/"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
